<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SCMultiomeCNRSR">
<title>Analysis of CITE-seq data with CiteFuse • SCMultiomeCNRSR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of CITE-seq data with CiteFuse">
<meta property="og:description" content="SCMultiomeCNRSR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SCMultiomeCNRSR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction_to_single-cell.html">Introduction to single-cell pre-processing</a>
    <a class="dropdown-item" href="../articles/basics_of_single-cell_annotation.html">Basics of single-cell annotation</a>
    <a class="dropdown-item" href="../articles/advancedphenotyping.html">Advanced Cell Phenotyping</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/carissaynchen/SCMultiome-CNRS-workshop-R">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<link href="citefuse_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="citefuse_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="citefuse_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="citefuse_files/datatables-binding-0.33/datatables.js"></script><link href="citefuse_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="citefuse_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="citefuse_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="citefuse_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="citefuse_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analysis of CITE-seq data with CiteFuse</h1>
                        <h4 data-toc-skip class="author">Carissa Chen<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:cchen@cmri.org.au" class="email"&gt;cchen@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a> Sharon
Long<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:slong@cmri.org.au" class="email"&gt;slong@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>
Pengyi Yang<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:pyang@cmri.org.au" class="email"&gt;pyang@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>3</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/carissaynchen/SCMultiome-CNRS-workshop-R/blob/HEAD/vignettes/citefuse.Rmd" class="external-link"><code>vignettes/citefuse.Rmd</code></a></small>
      <div class="d-none name"><code>citefuse.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>scRNA-seq has allowed us to dissect complex cell populations at the
transcriptomic level, however it lacks additional phenotypic information
such as the quantification of cell-surface proteins which is often
useful to distinguish closely related cell states (eg. in immunology).
CITE-seq uses oligonucleotide-labeled antibodies to integrate cellular
protein and transcriptomic measurements into a single-cell readout. In
this first section where we introduce single-cell multi-modal analysis,
we will learn how to pre-process and integratively explore CITE-seq data
using the <code>CiteFuse</code> package.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="preparation-and-assumed-knowledge">Preparation and assumed knowledge<a class="anchor" aria-label="anchor" href="#preparation-and-assumed-knowledge"></a>
</h3>
<ul>
<li><a href="https://www.nature.com/articles/nmeth.4380" class="external-link">CITE-seq
technology</a></li>
<li><a href="https://doi.org/10.1093/bioinformatics/btaa282" class="external-link">Analysing
CITE-seq with CiteFuse</a></li>
<li><a href="https://sydneybiox.github.io/CiteFuse/" class="external-link">CiteFuse
vignette</a></li>
</ul>
<p><br></p>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>understand the output of CITE-seq data</li>
<li>how to pre-process CITE-seq data</li>
<li>perform exploratory analysis and learn how to jointly visualise
modalities</li>
<li>integrate RNA-ADT information to construct interaction networks</li>
</ul>
<p><br></p>
</div>
<div class="section level3">
<h3 id="schedule">Schedule<a class="anchor" aria-label="anchor" href="#schedule"></a>
</h3>
<p>Structure of this section:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction to <code>CiteFuse</code>
</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Performing doublet detection</td>
<td>Xm</td>
</tr>
<tr class="odd">
<td>Integration and clustering of modalities</td>
<td>Xm</td>
</tr>
<tr class="even">
<td>Exploratory analysis and differential gene detection</td>
<td>Xm</td>
</tr>
<tr class="odd">
<td>Integrative analysis of RNA and ADT modalities</td>
<td>Xm</td>
</tr>
<tr class="even">
<td>Concluding remarks</td>
<td></td>
</tr>
</tbody>
</table>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The CITE-seq protocol involves conjugating antibodies to
oligonucleotides, where each antibody is specific to a cell surface
protein marker. The antibodies bind to their specific protein marker,
and mRNA is lysed from the cell. Both molecular species are captured for
scRNA-sequencing. In a process called ‘cell hashing’, cells are stained
with sample-specific oligonucleotide-labelled antibodies (‘hashing
antibodies’), thus allowing the samples to be multiplexed and sequenced
in a single experiment and later traced back to their origin. Thus,
allowing for high-throughput and high-dimensional multi-omic single-cell
analysis.</p>
<p><code>CiteFuse</code> is a computational framework that implements a
suite of methods and tools for CITE-seq data from pre-processing through
to integrative analytics. This includes doublet detection, network-based
modality integration, cell type clustering, differential RNA and protein
expression analysis, ADT evaluation, ligand-receptor interaction
analysis, and interactive web-based visualisation of the analyses.</p>
<p>In this workshop, we demonstrate the usage of <code>CiteFuse</code>
on a subset data of CITE-seq data from human PBMCs as an example <a href="https://doi.org/10.1038/s41592-019-0392-0" class="external-link">Mimitou et al.,
2019</a>.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="loading-libraries-and-the-data">Loading libraries and the data<a class="anchor" aria-label="anchor" href="#loading-libraries-and-the-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CiteFuse</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/DT" class="external-link">DT</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mclust-org.github.io/mclust/" class="external-link">mclust</a></span><span class="op">)</span></span></code></pre></div>
<p>Here, we start from a list of three matrices of unique molecular
identifier (UMI), antibody derived tags (ADT) and hashtag
oligonucleotide (HTO) count, which have common cell names.</p>
<p>There are 500 cells in our subsetted dataset. And characteristically
of CITE-seq data, the matrices are matched, meaning that for any given
cell we know the expression level of their RNA transcripts (genome-wide)
and its corresponding cell surface protein expression. The
<code>preprocessing</code> function will utilise the three matrices and
its common cell names to create a <code>SingleCellExperiment</code>
object, which stores RNA data in an <code>assay</code> and
<code>ADT</code> and <code>HTO</code> data within in the
<code>altExp</code> slot.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"CITEseq_example"</span>, package <span class="op">=</span> <span class="st">"CiteFuse"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">CITEseq_example</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RNA" "ADT" "HTO"</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">CITEseq_example</span>, <span class="va">dim</span><span class="op">)</span></span>
<span><span class="co">#&gt; $RNA</span></span>
<span><span class="co">#&gt; [1] 19521   500</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ADT</span></span>
<span><span class="co">#&gt; [1]  49 500</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $HTO</span></span>
<span><span class="co">#&gt; [1]   4 500</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/preprocessing.html" class="external-link">preprocessing</a></span><span class="op">(</span><span class="va">CITEseq_example</span><span class="op">)</span></span>
<span><span class="va">sce_citeseq</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 19521 500 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(19521): hg19_AL627309.1 hg19_AL669831.5 ... hg19_MT-ND6</span></span>
<span><span class="co">#&gt;   hg19_MT-CYB</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(500): AAGCCGCGTTGTCTTT GATCGCGGTTATCGGT ... TTGGCAACACTAGTAC</span></span>
<span><span class="co">#&gt;   GCTGCGAGTTGTGGCC</span></span>
<span><span class="co">#&gt; colData names(0):</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(2): ADT HTO</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="performing-doublet-detection">Performing doublet detection<a class="anchor" aria-label="anchor" href="#performing-doublet-detection"></a>
</h2>
<p>An important step in single cell data analysis is the removal of
doublets. Doublets form as a result of co-encapsulation of cells within
a droplet, leading to a hybrid transcriptome from two or more cells. In
<code>CiteFuse</code>, we implement a step-wise doublet detection
approach to remove doublets. We first identify the cross-sample
doublets, where cells from different samples are captured as a ‘single’
cell. Then, we identify within-sample doublets, where two cells from the
same sample are captured together in a droplet.</p>
<div class="section level3">
<h3 id="normalisation-and-initial-visualisation-of-hto-counts">Normalisation and initial visualisation of HTO counts<a class="anchor" aria-label="anchor" href="#normalisation-and-initial-visualisation-of-hto-counts"></a>
</h3>
<p>The function <code>normaliseExprs</code> is used to scale the
alternative expression. Here, we used it to perform log-transformation
of the <code>HTO</code> count, by setting
<code>transform = "log"</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html" class="external-link">normaliseExprs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">sce_citeseq</span>, </span>
<span>                              altExp_name <span class="op">=</span> <span class="st">"HTO"</span>, </span>
<span>                              transform <span class="op">=</span> <span class="st">"log"</span><span class="op">)</span></span></code></pre></div>
<p>Then we can perform dimension reduction on the <code>HTO</code> count
by using <code>runTSNE</code> or <code>runUMAP</code>, then use
<code>visualiseDim</code> function to visualise the reduced dimension
plot. Our CITE-seq dataset contain data from four samples that were
pooled before sequencing. The samples were multiplexed through cell
hashing <a href="https://doi.org/10.1038/nmeth.4380" class="external-link">Stoeckius et al.,
2018</a>. The four clusters observed on reduced dimension plots equate
to the four different samples.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>                               altexp <span class="op">=</span> <span class="st">"HTO"</span>, </span>
<span>                               name <span class="op">=</span> <span class="st">"TSNE_HTO"</span>, </span>
<span>                               pca <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>                               altexp <span class="op">=</span> <span class="st">"HTO"</span>, </span>
<span>                               name <span class="op">=</span> <span class="st">"UMAP_HTO"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>             dimNames <span class="op">=</span> <span class="st">"TSNE_HTO"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"tSNE (HTO)"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>             dimNames <span class="op">=</span> <span class="st">"UMAP_HTO"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"UMAP (HTO)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="doublet-identification-step-1-cross-sample-doublet-detection">Doublet identification step 1: cross-sample doublet detection<a class="anchor" aria-label="anchor" href="#doublet-identification-step-1-cross-sample-doublet-detection"></a>
</h3>
<p>To predict cross-sample doublets, we fit a two-component Gaussian
mixture model to each log-transformed HTO count. The intersection point
defined from the mixture model is used to categorize each cell in terms
of whether the HTO signal is either high or low. Because a sample is
stained with a hashtag that is distinctive to that sample, all cells
should in theory be associated with only one HTO signal and show
negligible signal of others. Therefore, any cells found to have a single
high-signal HTO are considered singlets whilst those that have two or
more high-signal HTOs are considered doublets or multiplets. Cells
without any high-signal HTOs are considered as empty droplets.</p>
<p>We run <code>crossSampleDoublets</code> where results of the cross
sample doublets are then saved in the <code>colData</code> slot as
<code>doubletClassify_between_label</code> and
<code>doubletClassify_between_class</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/crossSampleDoublets.html" class="external-link">crossSampleDoublets</a></span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span></span>
<span><span class="co">#&gt; number of iterations= 20 </span></span>
<span><span class="co">#&gt; number of iterations= 24 </span></span>
<span><span class="co">#&gt; number of iterations= 46 </span></span>
<span><span class="co">#&gt; number of iterations= 50</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">doubletClassify_between_label</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                 1                 2                 3                 4 </span></span>
<span><span class="co">#&gt;               115               121                92               129 </span></span>
<span><span class="co">#&gt; doublet/multiplet </span></span>
<span><span class="co">#&gt;                43</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">doubletClassify_between_class</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; doublet/multiplet           Singlet </span></span>
<span><span class="co">#&gt;                43               457</span></span></code></pre></div>
<p>We can then highlight the cross-sample doublets in our tSNE plot of
HTO count. How would we visualise the between label and between class
doublets side-by-side?</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>             dimNames <span class="op">=</span> <span class="st">"TSNE_HTO"</span>, </span>
<span>             colour_by <span class="op">=</span> <span class="st">"doubletClassify_between_label"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-8-1.png" width="576"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># insert code to visualise "doubletClassify_between_class"</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="doublet-identification-step-2-within-sample-doublet-detection">Doublet identification step 2: within-sample doublet detection<a class="anchor" aria-label="anchor" href="#doublet-identification-step-2-within-sample-doublet-detection"></a>
</h3>
<p>Data filtered by cross-sample doublets are next subject to
within-sample doublet identification using a density-based spatial
clustering and noise detection algorithm (DBSCAN) on a matrix comprising
of two types of features—total unique molecular identifier (UMI) count
and log-transformed HTO count. The two parameters used in the DBSCAN for
this study are eps=190 and minPts=50. The smallest cluster
(i.e. outliers) from DBSCAN clustering is assigned as the within-sample
doublets.</p>
<p>We then identify the within-sample doublets via the
<code>withinSampleDoublets</code> function. The results of the cross
sample doublets are then saved in the <code>colData</code> as
<code>doubletClassify_within_label</code> and
<code>doubletClassify_within_class</code>.</p>
<p><strong>Questions:</strong></p>
<ul>
<li>What are the <code>eps</code> and <code>minPts</code>
parameters?</li>
<li>What happens when you vary these parameters?</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/withinSampleDoublets.html" class="external-link">withinSampleDoublets</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, minPts <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">doubletClassify_within_label</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Doublets(Within)_1  Doublets(Within)_2  Doublets(Within)_3  Doublets(Within)_4 </span></span>
<span><span class="co">#&gt;                   3                   7                   4                   6 </span></span>
<span><span class="co">#&gt; NotDoublets(Within) </span></span>
<span><span class="co">#&gt;                 480</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">doubletClassify_within_class</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Doublet Singlet </span></span>
<span><span class="co">#&gt;      20     480</span></span></code></pre></div>
<p>Again, we can visualise the within-sample doublets in our tSNE plot.
How would we visualise the within label and within class doublets
side-by-side?</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>             dimNames <span class="op">=</span> <span class="st">"TSNE_HTO"</span>, </span>
<span>             colour_by <span class="op">=</span> <span class="st">"doubletClassify_within_label"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-11-1.png" width="576"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># insert code to visualise "doubletClassify_within_class"</span></span></code></pre></div>
<p>Furthermore, <code>plotHTO</code> function allows us to plot the
pairwise scatter HTO count. Any cells that show co-expression of
orthogonal HTOs (red) are considered as doublets. Which comparison has
the most cross-sample doublets?</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/plotHTO.html" class="external-link">plotHTO</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-12-1.png" width="576"></p>
<p>Finally, we can filter out the doublet cells (both within and between
batches) for the downstream analysis.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="va">sce_citeseq</span><span class="op">[</span>, <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">doubletClassify_within_class</span> <span class="op">==</span> <span class="st">"Singlet"</span> <span class="op">&amp;</span> <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">doubletClassify_between_class</span> <span class="op">==</span> <span class="st">"Singlet"</span><span class="op">]</span></span>
<span><span class="va">sce_citeseq</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 19521 437 </span></span>
<span><span class="co">#&gt; metadata(3): doubletClassify_between_threshold</span></span>
<span><span class="co">#&gt;   doubletClassify_between_resultsMat doubletClassify_within_resultsMat</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(19521): hg19_AL627309.1 hg19_AL669831.5 ... hg19_MT-ND6</span></span>
<span><span class="co">#&gt;   hg19_MT-CYB</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(437): GATCGCGGTTATCGGT GGCTGGTAGAGGTTAT ... TTGGCAACACTAGTAC</span></span>
<span><span class="co">#&gt;   GCTGCGAGTTGTGGCC</span></span>
<span><span class="co">#&gt; colData names(5): doubletClassify_between_label</span></span>
<span><span class="co">#&gt;   doubletClassify_between_class nUMI doubletClassify_within_label</span></span>
<span><span class="co">#&gt;   doubletClassify_within_class</span></span>
<span><span class="co">#&gt; reducedDimNames(2): TSNE_HTO UMAP_HTO</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(2): ADT HTO</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="integration-of-rna-and-adt-using-snf">Integration of RNA and ADT using <code>SNF</code><a class="anchor" aria-label="anchor" href="#integration-of-rna-and-adt-using-snf"></a>
</h2>
<p>The next step of analysis is to integrate the RNA and ADT matrix. We
use a popular integration algorithm called similarity network fusion
(SNF) to integrate the multi-omic data.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html" class="external-link">normaliseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, altExp_name <span class="op">=</span> <span class="st">"ADT"</span>, transform <span class="op">=</span> <span class="st">"log"</span><span class="op">)</span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/CiteFuse.html" class="external-link">CiteFuse</a></span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating affinity matrix </span></span>
<span><span class="co">#&gt; Performing SNF</span></span></code></pre></div>
<p>We now proceed with the fused matrix, which is stored as
<code>SNF_W</code> in the metadata slot of our <code>sce_citeseq</code>
object. We can calculate the t-SNE to generate an integrated embedding
using the SNF network.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/reducedDimSNF.html" class="external-link">reducedDimSNF</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                             method <span class="op">=</span> <span class="st">"tSNE"</span>, </span>
<span>                             dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="how-many-clusters-are-in-my-data">How many clusters are in my data?<a class="anchor" aria-label="anchor" href="#how-many-clusters-are-in-my-data"></a>
</h2>
<div class="section level3">
<h3 id="clustering-using-louvain-clustering">Clustering using Louvain clustering<a class="anchor" aria-label="anchor" href="#clustering-using-louvain-clustering"></a>
</h3>
<p>As well as spectral clustering, CiteFuse can implement Louvain
clustering if users wish to use another clustering method. We use the
<code>igraph</code> package, and any community detection algorithms
available in their package can be selected by changing the
<code>method</code> parameter.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="va">SNF_W_louvain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/igraphClustering.html" class="external-link">igraphClustering</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, method <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">SNF_W_louvain</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNF_W_louvain</span></span>
<span><span class="co">#&gt;   1   2   3   4   5   6   7 </span></span>
<span><span class="co">#&gt;  88 137  65  32  51  28  36</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_louvain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">SNF_W_louvain</span><span class="op">)</span></span></code></pre></div>
<p>The outcome of the clustering can be easily visualised on a reduced
dimensions plot by highlighting the points by cluster label.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span>, colour_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"tSNE (SNF louvain clustering)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-17-1.png" width="1152"></p>
<p>Alternatively, the clustering can be easily visualised using a KNN
graph labelled by cluster number.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseKNN.html" class="external-link">visualiseKNN</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, colour_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="optional-clustering-using-spectral-clustering">[Optional] Clustering using Spectral clustering<a class="anchor" aria-label="anchor" href="#optional-clustering-using-spectral-clustering"></a>
</h3>
<p>CiteFuse implements two different clustering algorithms on the fused
matrix, spectral clustering and Louvain clustering. First, we perform
spectral clustering with sufficient numbers of <code>K</code> and use
the eigen values to determine the optimal number of clusters. What is
the optimal number of clusters here?</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SNF_W_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html" class="external-link">spectralClustering</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"SNF_W"</span><span class="op">]</span><span class="op">]</span>, K <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing Spectral Clustering</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">SNF_W_clust</span><span class="op">$</span><span class="va">eigen_values</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">SNF_W_clust</span><span class="op">$</span><span class="va">eigen_values</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 5</span></span></code></pre></div>
<p>Using the optimal cluster number defined from the previous step, we
can now use the <code>spectralClustering</code> function to cluster the
single cells by specifying the number of clusters in <code>K</code>. The
function takes a cell-to-cell similarity matrix as an input. We have
already created the fused similarity matrix from <code>CiteFuse</code>.
Since the <code>CiteFuse</code> function creates and stores the
similarity matries from ADT and RNA expression, as well the fused
matrix, we can use these two to compare the clustering outcomes by data
modality.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SNF_W_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html" class="external-link">spectralClustering</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"SNF_W"</span><span class="op">]</span><span class="op">]</span>, K <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing Spectral Clustering</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">SNF_W_clust</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SNF_W1_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html" class="external-link">spectralClustering</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"ADT_W"</span><span class="op">]</span><span class="op">]</span>, K <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing Spectral Clustering</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">ADT_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">SNF_W1_clust</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SNF_W2_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html" class="external-link">spectralClustering</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"RNA_W"</span><span class="op">]</span><span class="op">]</span>, K <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing Spectral Clustering</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">RNA_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">SNF_W2_clust</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="exploratory-analysis-of-feature-expression-across-modalities">Exploratory analysis of feature expression across modalities<a class="anchor" aria-label="anchor" href="#exploratory-analysis-of-feature-expression-across-modalities"></a>
</h2>
<p>CiteFuse has a wide range of visualisation tools to facilitate
exploratory analysis of CITE-seq data. The <code>visualiseExprs</code>
function is an easy-to-use function to generate boxplots, violin plots,
jitter plots, density plots, and pairwise scatter/density plots of genes
and proteins expressed in the data. The plots can be grouped by using
the cluster labels stored in the <code>sce_citeseq</code> object.</p>
<p>The expression of genes and proteins can be visualised by changing
the <code>colour_by</code> parameter to assess the clusters. As an
example, we highlight the plot by the RNA and ADT expression level of
CD8.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span>, </span>
<span>                   colour_by <span class="op">=</span> <span class="st">"hg19_CD8A"</span>,</span>
<span>                   data_from <span class="op">=</span> <span class="st">"assay"</span>,</span>
<span>                   assay_name <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"tSNE: hg19_CD8A (RNA expression)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span>, </span>
<span>                   colour_by <span class="op">=</span> <span class="st">"CD8"</span>,</span>
<span>                   data_from <span class="op">=</span> <span class="st">"altExp"</span>,</span>
<span>                   altExp_assay_name <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"tSNE: CD8 (ADT expression)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g1</span> <span class="op">+</span> <span class="va">g2</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-22-1.png" width="1152"></p>
<p><strong>Alternative visualisation 1:</strong> We can plot a boxplot
of the expression of a subset of genes per cluster detected by the
louvain clustering algorithm.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html" class="external-link">visualiseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>               plot <span class="op">=</span> <span class="st">"boxplot"</span>, </span>
<span>               altExp_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>               group_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span>,</span>
<span>               feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-23-1.png" width="768"></p>
<p><strong>Alternative visualisation 2:</strong> We can plot a violin
plot of the expression of a subset of genes per cluster detected by the
louvain clustering algorithm.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html" class="external-link">visualiseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>               plot <span class="op">=</span> <span class="st">"violin"</span>,</span>
<span>               altExp_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>               group_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span>,</span>
<span>               feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p><strong>Alternative visualisation 3:</strong> We can plot a jitter of
the expression of a subset of genes per cluster detected by the louvain
clustering algorithm.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html" class="external-link">visualiseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>               plot <span class="op">=</span> <span class="st">"jitter"</span>, </span>
<span>               altExp_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>               group_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span>,</span>
<span>               feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p><strong>Alternative visualisation 4:</strong> We can plot a density
plot of the expression of a subset of genes per cluster detected by the
louvain clustering algorithm.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html" class="external-link">visualiseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>               plot <span class="op">=</span> <span class="st">"density"</span>, </span>
<span>               altExp_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>               group_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span>,</span>
<span>               feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<p><strong>Questions:</strong></p>
<ul>
<li>From these collection of visualisations, do you notice any markers
that are differentially expressed in any clusters?</li>
<li>How would you plot the above, but for ADT?</li>
</ul>
<p><strong>Alternative visualisation 5:</strong> We can also create a
pairwise scatter plot to visualise the distribution of cells that are
expressing certain markers.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html" class="external-link">visualiseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, altExp_name <span class="op">=</span> <span class="st">"ADT"</span>, </span>
<span>               plot <span class="op">=</span> <span class="st">"pairwise"</span>, </span>
<span>               feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; number of iterations= 19 </span></span>
<span><span class="co">#&gt; number of iterations= 26</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-27-1.png" width="768"></p>
<p><strong>Questions:</strong></p>
<ul>
<li>What do each of the quadrants correspond to?</li>
<li>Do we have many cells that co-express both markers?</li>
</ul>
<p><br></p>
</div>
<div class="section level2">
<h2 id="what-genes-are-differentially-expressed-between-my-clusters">What genes are differentially expressed between my clusters?<a class="anchor" aria-label="anchor" href="#what-genes-are-differentially-expressed-between-my-clusters"></a>
</h2>
<div class="section level3">
<h3 id="perform-de-analysis-with-wilcoxon-rank-sum-test">Perform DE Analysis with Wilcoxon Rank Sum test<a class="anchor" aria-label="anchor" href="#perform-de-analysis-with-wilcoxon-rank-sum-test"></a>
</h3>
<p><code>CiteFuse</code> can also calculates differentially expressed
(DE) genes through the <code>DEgenes</code> function. The cluster
grouping to use must be specified in the <code>group</code> parameter.
If <code>altExp_name</code> is not specified, RNA expression will be
used as the default expression matrix.</p>
<p>Results form the DE analysis is stored in <code>sce_citeseq</code> as
<code>DE_res_RNA_filter</code> and <code>DE_res_ADT_filter</code> for
RNA and ADT expression, respectively.</p>
<div class="section level4">
<h4 id="for-rna-modality">For RNA modality<a class="anchor" aria-label="anchor" href="#for-rna-modality"></a>
</h4>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DE will be performed for RNA if altExp_name = "none" </span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEgenes.html" class="external-link">DEgenes</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                       altExp_name <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                       group <span class="op">=</span> <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_louvain</span>,</span>
<span>                       return_all <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       exprs_pct <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/selectDEgenes.html" class="external-link">selectDEgenes</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                             altExp_name <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"DE_res_RNA_filter"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                 digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["1.hg19_LTB","1.hg19_TCF7","1.hg19_IL7R","1.hg19_CD27","1.hg19_LDHB","1.hg19_LEF1","1.hg19_SELL","1.hg19_NPM1","1.hg19_NOSIP","1.hg19_CAMK4","2.hg19_CD8A","2.hg19_CCL5","2.hg19_GZMH","2.hg19_ZNF683","2.hg19_GNLY","2.hg19_FGFBP2","2.hg19_KLRD1","2.hg19_GZMB","2.hg19_CST7","2.hg19_NKG7","3.hg19_CD8B","3.hg19_CD8A","3.hg19_CNN2","3.hg19_GYPC","3.hg19_MGAT4A","3.hg19_AL136454.1","3.hg19_ATM","3.hg19_CD7","3.hg19_IL7R","3.hg19_SNHG8","4.hg19_IL7R","4.hg19_S100A4","4.hg19_S100A10","4.hg19_IL32","4.hg19_RORA","4.hg19_FAM107B","4.hg19_ITGB1","4.hg19_EML4","4.hg19_CD6","4.hg19_CD44","5.hg19_SH2D1B","5.hg19_KLRF1","5.hg19_FCER1G","5.hg19_MYOM2","5.hg19_CLIC3","5.hg19_TRDC","5.hg19_TYROBP","5.hg19_SPON2","5.hg19_IL2RB","5.hg19_KLRB1","6.hg19_GZMK","6.hg19_KLRB1","6.hg19_CEBPD","6.hg19_NCR3","6.hg19_IL7R","6.hg19_DUSP2","6.hg19_KLRG1","6.hg19_CD69","6.hg19_SATB1","6.hg19_LYAR","7.hg19_FCN1","7.hg19_LYZ","7.hg19_SPI1","7.hg19_CYBB","7.hg19_S100A9","7.hg19_HLA-DRA","7.hg19_CD68","7.hg19_S100A8","7.hg19_CSTA","7.hg19_VCAN"],[" 3778"," 5778"," 5710"," 8479"," 5946"," 8339"," 8482"," 7150"," 8192"," 9504"," 5747"," 5448"," 6474","10888"," 7146"," 8356"," 8427"," 8318"," 7688"," 7805"," 5202"," 6890"," 8274"," 8416"," 9054"," 9084"," 9446"," 9258"," 9414"," 9625"," 2594"," 3374"," 3682"," 3807"," 4564"," 4536"," 4478"," 4754"," 4726"," 4655"," 3622"," 2767"," 3297"," 2003"," 3493"," 4077"," 2702"," 2827"," 3138"," 3175"," 2469","  812"," 2946"," 3161"," 2256"," 2420"," 2906"," 2862"," 3532"," 3581"," 2206"," 2211"," 2214"," 2406"," 2614","  669"," 3008"," 2630"," 3208"," 3208"],["4.9e-34","3.2e-29","3.2e-23","2.3e-21","2.7e-20","5.2e-19","2.9e-15","6.2e-15","1.5e-14","1.6e-13","1.4e-39","1.9e-36","7.0e-36","6.6e-32","2.8e-29","3.3e-28","2.0e-27","3.7e-27","4.3e-27","2.1e-26","5.1e-21","1.6e-09","6.3e-06","1.4e-05","1.2e-04","1.1e-03","1.7e-03","1.9e-03","1.9e-03","2.2e-03","7.5e-10","6.3e-06","4.2e-05","9.3e-05","1.1e-03","1.4e-03","1.6e-03","2.8e-03","2.9e-03","3.5e-03","8.2e-50","7.0e-36","1.9e-34","4.7e-33","1.6e-32","1.1e-31","6.2e-27","1.2e-26","6.4e-25","2.6e-23","6.9e-23","7.3e-22","2.8e-11","4.1e-10","5.0e-09","5.5e-09","2.2e-07","3.5e-07","3.1e-06","1.4e-04","5.8e-66","1.7e-63","2.0e-63","2.9e-63","4.6e-58","6.3e-58","3.3e-55","1.3e-53","1.5e-52","1.5e-52"],["4.9e-34","3.2e-29","3.2e-23","2.3e-21","2.7e-20","5.2e-19","2.9e-15","6.2e-15","1.5e-14","1.6e-13","1.4e-39","1.9e-36","7.0e-36","6.6e-32","2.8e-29","3.3e-28","2.0e-27","3.7e-27","4.3e-27","2.1e-26","5.1e-21","1.6e-09","6.3e-06","1.4e-05","1.2e-04","1.1e-03","1.7e-03","1.9e-03","1.9e-03","2.2e-03","7.5e-10","6.3e-06","4.2e-05","9.3e-05","1.1e-03","1.4e-03","1.6e-03","2.8e-03","2.9e-03","3.5e-03","8.2e-50","7.0e-36","1.9e-34","4.7e-33","1.6e-32","1.1e-31","6.2e-27","1.2e-26","6.4e-25","2.6e-23","6.9e-23","7.3e-22","2.8e-11","4.1e-10","5.0e-09","5.5e-09","2.2e-07","3.5e-07","3.1e-06","1.4e-04","5.8e-66","1.7e-63","2.0e-63","2.9e-63","4.6e-58","6.3e-58","3.3e-55","1.3e-53","1.5e-52","1.5e-52"],["0.5647","0.2360","0.7507","0.1172","0.7219","0.1807","0.3341","1.2433","0.3780","0.1817","0.4483","1.4298","0.5045","0.0533","1.8054","0.4782","0.4813","0.6659","1.1052","1.6509","0.2898","0.7847","0.4991","0.4664","0.3319","0.9119","0.4719","0.9524","0.9702","0.3864","0.9345","2.3209","1.2816","1.7825","0.4297","0.4573","0.6891","0.3681","0.4142","0.5453","0.0208","0.1328","0.1807","0.2513","0.0989","0.0939","0.4674","0.2817","0.2249","0.4483","0.1128","0.4798","0.2070","0.1678","0.9438","0.5254","0.4771","0.4036","0.2122","0.4926","0.0000","0.0016","0.0027","0.0000","0.0020","0.0574","0.0000","0.0079","0.0000","0.0000"],["2.39","1.19","2.18","0.64","1.78","0.82","1.04","2.07","1.03","0.62","1.91","3.77","1.97","0.79","4.59","1.67","1.57","2.02","2.64","3.82","1.31","1.60","0.93","0.83","0.62","1.27","0.75","1.30","1.43","0.64","2.36","3.31","1.96","2.63","0.84","0.80","1.18","0.77","0.74","0.93","0.91","1.22","1.74","1.83","1.40","0.98","2.12","1.78","1.34","1.99","0.91","2.80","1.00","0.83","2.43","1.56","1.22","1.08","0.63","1.08","1.88","2.82","1.17","0.83","2.60","2.87","0.76","2.89","1.07","1.54"],["0.2951","0.1748","0.3352","0.1003","0.5129","0.1461","0.2264","0.7221","0.3123","0.1633","0.2600","0.4867","0.2633","0.0333","0.4633","0.2533","0.2733","0.2900","0.4633","0.4933","0.1882","0.3844","0.3790","0.3817","0.2984","0.6371","0.3844","0.5860","0.4220","0.3253","0.4222","0.8938","0.7432","0.7654","0.3358","0.3802","0.4568","0.3160","0.3358","0.4222","0.0155","0.0959","0.0777","0.1788","0.0855","0.0596","0.1839","0.1917","0.1865","0.2021","0.0587","0.2323","0.1345","0.1296","0.4279","0.3545","0.3032","0.3472","0.1980","0.3545","0.0000","0.0025","0.0025","0.0000","0.0025","0.0474","0.0000","0.0075","0.0000","0.0000"],["0.93","0.77","0.95","0.55","0.85","0.59","0.66","0.95","0.70","0.55","0.90","1.00","0.90","0.50","0.99","0.82","0.83","0.88","0.99","1.00","0.72","0.89","0.68","0.68","0.55","0.83","0.60","0.77","0.68","0.52","0.94","1.00","0.97","0.94","0.59","0.62","0.69","0.53","0.59","0.69","0.65","0.80","0.75","0.88","0.69","0.65","0.94","0.80","0.78","0.86","0.64","0.96","0.61","0.57","0.93","0.86","0.82","0.79","0.57","0.68","0.69","0.69","0.69","0.67","0.64","0.92","0.58","0.64","0.56","0.56"],["1.82","0.95","1.43","0.52","1.06","0.64","0.71","0.83","0.65","0.44","1.46","2.34","1.47","0.74","2.79","1.19","1.09","1.35","1.54","2.16","1.02","0.82","0.43","0.36","0.29","0.36","0.28","0.35","0.46","0.26","1.42","0.99","0.68","0.85","0.41","0.34","0.49","0.40","0.32","0.38","0.89","1.08","1.56","1.58","1.30","0.89","1.66","1.49","1.11","1.54","0.80","2.32","0.80","0.66","1.48","1.03","0.74","0.68","0.41","0.58","1.88","2.81","1.17","0.83","2.59","2.81","0.76","2.88","1.07","1.54"],["0.64","0.60","0.62","0.45","0.34","0.44","0.43","0.23","0.39","0.38","0.64","0.51","0.63","0.47","0.53","0.57","0.56","0.59","0.53","0.51","0.53","0.51","0.30","0.30","0.26","0.19","0.22","0.18","0.25","0.20","0.52","0.11","0.23","0.17","0.26","0.24","0.23","0.22","0.26","0.27","0.63","0.71","0.67","0.70","0.60","0.59","0.76","0.61","0.60","0.66","0.58","0.73","0.47","0.44","0.50","0.50","0.52","0.44","0.37","0.32","0.69","0.69","0.69","0.67","0.64","0.87","0.58","0.63","0.56","0.56"],["hg19_LTB","hg19_TCF7","hg19_IL7R","hg19_CD27","hg19_LDHB","hg19_LEF1","hg19_SELL","hg19_NPM1","hg19_NOSIP","hg19_CAMK4","hg19_CD8A","hg19_CCL5","hg19_GZMH","hg19_ZNF683","hg19_GNLY","hg19_FGFBP2","hg19_KLRD1","hg19_GZMB","hg19_CST7","hg19_NKG7","hg19_CD8B","hg19_CD8A","hg19_CNN2","hg19_GYPC","hg19_MGAT4A","hg19_AL136454.1","hg19_ATM","hg19_CD7","hg19_IL7R","hg19_SNHG8","hg19_IL7R","hg19_S100A4","hg19_S100A10","hg19_IL32","hg19_RORA","hg19_FAM107B","hg19_ITGB1","hg19_EML4","hg19_CD6","hg19_CD44","hg19_SH2D1B","hg19_KLRF1","hg19_FCER1G","hg19_MYOM2","hg19_CLIC3","hg19_TRDC","hg19_TYROBP","hg19_SPON2","hg19_IL2RB","hg19_KLRB1","hg19_GZMK","hg19_KLRB1","hg19_CEBPD","hg19_NCR3","hg19_IL7R","hg19_DUSP2","hg19_KLRG1","hg19_CD69","hg19_SATB1","hg19_LYAR","hg19_FCN1","hg19_LYZ","hg19_SPI1","hg19_CYBB","hg19_S100A9","hg19_HLA-DRA","hg19_CD68","hg19_S100A8","hg19_CSTA","hg19_VCAN"],["1","1","1","1","1","1","1","1","1","1","2","2","2","2","2","2","2","2","2","2","3","3","3","3","3","3","3","3","3","3","4","4","4","4","4","4","4","4","4","4","5","5","5","5","5","5","5","5","5","5","6","6","6","6","6","6","6","6","6","6","7","7","7","7","7","7","7","7","7","7"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stats.W<\/th>\n      <th>pval<\/th>\n      <th>p.adjust<\/th>\n      <th>meanExprs.1<\/th>\n      <th>meanExprs.2<\/th>\n      <th>meanPct.1<\/th>\n      <th>meanPct.2<\/th>\n      <th>meanDiff<\/th>\n      <th>pctDiff<\/th>\n      <th>name<\/th>\n      <th>group<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"stats.W","targets":1},{"name":"pval","targets":2},{"name":"p.adjust","targets":3},{"name":"meanExprs.1","targets":4},{"name":"meanExprs.2","targets":5},{"name":"meanPct.1","targets":6},{"name":"meanPct.2","targets":7},{"name":"meanDiff","targets":8},{"name":"pctDiff","targets":9},{"name":"name","targets":10},{"name":"group","targets":11}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level4">
<h4 id="for-adt-modality">For ADT modality<a class="anchor" aria-label="anchor" href="#for-adt-modality"></a>
</h4>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEgenes.html" class="external-link">DEgenes</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                       altExp_name <span class="op">=</span> <span class="st">"ADT"</span>, </span>
<span>                       group <span class="op">=</span> <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_louvain</span>,</span>
<span>                       return_all <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       exprs_pct <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/selectDEgenes.html" class="external-link">selectDEgenes</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                             altExp_name <span class="op">=</span> <span class="st">"ADT"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"DE_res_ADT_filter"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                 digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"filter":"none","vertical":false,"data":[["1.CD28","1.CD26 (Adenosine)","1.CD3","5","6.CD26 (Adenosine)","6.CD28","7.LAMP1","7.CD45RO","7.CD1a","7.CD45","7.B7-H1 (PD-L1)","7.B220 (CD45R)","7.PD-L1 (CD274)","7.CTLA4","7.CD366 (tim3)","7.CD14"],[" 4060","10844","10826"," 3109"," 2451"," 2897"," 1550"," 1571"," 1774"," 1958"," 2570"," 2354"," 2302"," 2336"," 2611"," 3148"],["8.1e-27","2.2e-06","1.7e-05","1.0e-16","1.8e-08","1.1e-05","2.2e-15","3.8e-15","2.7e-14","3.4e-14","3.1e-12","6.0e-12","7.3e-12","1.2e-11","3.4e-11","4.6e-11"],["8.1e-27","2.2e-06","1.7e-05","1.0e-16","1.8e-08","1.1e-05","2.2e-15","3.8e-15","2.7e-14","3.4e-14","3.1e-12","6.0e-12","7.3e-12","1.2e-11","3.4e-11","4.6e-11"],["1.28","0.45","1.43","0.65","0.46","1.49","0.92","1.14","1.09","0.56","0.41","0.79","1.13","1.22","0.66","0.27"],["2.54","0.86","1.82","1.74","1.65","2.29","2.58","2.62","2.28","1.50","1.36","1.72","2.20","2.27","2.14","1.61"],["0.83","0.38","0.85","0.54","0.41","0.85","0.77","0.85","0.89","0.58","0.46","0.75","0.87","0.87","0.56","0.33"],["0.99","0.62","0.97","0.94","0.75","0.96","1.00","1.00","1.00","0.97","0.86","0.97","0.97","0.97","0.86","0.69"],["1.26","0.40","0.39","1.09","1.20","0.80","1.66","1.49","1.19","0.93","0.95","0.93","1.08","1.05","1.49","1.34"],["0.16","0.24","0.11","0.40","0.34","0.11","0.23","0.15","0.11","0.39","0.40","0.22","0.10","0.10","0.30","0.37"],["CD28","CD26 (Adenosine)","CD3","CD366 (tim3)","CD26 (Adenosine)","CD28","LAMP1","CD45RO","CD1a","CD45","B7-H1 (PD-L1)","B220 (CD45R)","PD-L1 (CD274)","CTLA4","CD366 (tim3)","CD14"],["1","1","1","5","6","6","7","7","7","7","7","7","7","7","7","7"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stats.W<\/th>\n      <th>pval<\/th>\n      <th>p.adjust<\/th>\n      <th>meanExprs.1<\/th>\n      <th>meanExprs.2<\/th>\n      <th>meanPct.1<\/th>\n      <th>meanPct.2<\/th>\n      <th>meanDiff<\/th>\n      <th>pctDiff<\/th>\n      <th>name<\/th>\n      <th>group<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"stats.W","targets":1},{"name":"pval","targets":2},{"name":"p.adjust","targets":3},{"name":"meanExprs.1","targets":4},{"name":"meanExprs.2","targets":5},{"name":"meanPct.1","targets":6},{"name":"meanPct.2","targets":7},{"name":"meanDiff","targets":8},{"name":"pctDiff","targets":9},{"name":"name","targets":10},{"name":"group","targets":11}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="visualising-de-results">Visualising DE Results<a class="anchor" aria-label="anchor" href="#visualising-de-results"></a>
</h3>
<p>After running DE analysis, we obtain the p-value and relevant
statistics to rank our differentially expressed genes in each cluster.
To visualise a selection of top DE genes, we can use
`<code>DEcomparisonPlot</code>. This lollipop plot shows the -log10
transformed adjusted p-values of the specified genes in RNA and ADT in
each cluster.</p>
<p>It requires a list of features from both RNA and ADT that we are
interested in, as well as the output from the DE analysis.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_HLA-DRB1"</span>, <span class="st">"hg19_ITGAX"</span>, <span class="st">"hg19_NCAM1"</span>, <span class="st">"hg19_CD27"</span>, <span class="st">"hg19_CD19"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">adt_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"MHCII (HLA-DR)"</span>, <span class="st">"CD11c"</span>, <span class="st">"CD56"</span>, <span class="st">"CD27"</span>, <span class="st">"CD19"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rna_DEgenes_all</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"DE_res_RNA"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">adt_DEgenes_all</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"DE_res_ADT"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">feature_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>RNA <span class="op">=</span> <span class="va">rna_list</span>, ADT <span class="op">=</span> <span class="va">adt_list</span><span class="op">)</span></span>
<span><span class="va">de_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>RNA <span class="op">=</span> <span class="va">rna_DEgenes_all</span>, ADT <span class="op">=</span> <span class="va">adt_DEgenes_all</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEcomparisonPlot.html" class="external-link">DEcomparisonPlot</a></span><span class="op">(</span>de_list <span class="op">=</span> <span class="va">de_list</span>,</span>
<span>                 feature_list <span class="op">=</span> <span class="va">feature_list</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-30-1.png" width="768"></p>
<p><strong>Interpretation:</strong></p>
<p>For each louvain-generated groups, we compare the -log10 transformed
p-values of RNA genes (right) and ADT markers (left). The larger the
value, the more significant the p-value. From this, we can see that in
cluster 7, there is the significant differential expression of MHCII and
CD19. This suggests that cluster 7 are B cells, possibly a B cell
subtype.</p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="how-do-i-analyse-the-rna-and-adt-modalities-together">How do I analyse the RNA and ADT modalities together?<a class="anchor" aria-label="anchor" href="#how-do-i-analyse-the-rna-and-adt-modalities-together"></a>
</h2>
<div class="section level3">
<h3 id="is-there-a-relationship-between-important-molecular-factors-in-rna-and-adt">Is there a relationship between important molecular factors in RNA
and ADT?<a class="anchor" aria-label="anchor" href="#is-there-a-relationship-between-important-molecular-factors-in-rna-and-adt"></a>
</h3>
<p>We can generate a network diagram using the
<code>geneADTnetwork</code> function to examine the association between
the differentially expressed genes and protein markers. We specify the
unique DE genes across all clusters from the RNA and ADT modalities and
calculate the correlation between the genes using all cells. The nodes
denote proteins and RNA whilst the edges denote positive and negative
correlation in expression.</p>
<p><strong>Questions:</strong></p>
<ul>
<li>Do the results change if we use a different correlation method?</li>
<li>What if we vary the <code>cor_threshold</code> parameter?</li>
<li>How do we examine the network for a specific cluster?</li>
</ul>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RNA_feature_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">rna_DEgenes_all</span>, <span class="st">"[["</span>, <span class="st">"name"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ADT_feature_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">adt_DEgenes_all</span>, <span class="st">"[["</span>, <span class="st">"name"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="va">netw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/geneADTnetwork.html" class="external-link">geneADTnetwork</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>               RNA_exprs_value <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>               altExp_name <span class="op">=</span> <span class="st">"ADT"</span>,</span>
<span>               altExp_exprs_value <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>               cell_subset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>               cor_threshold <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>               RNA_feature_subset <span class="op">=</span> <span class="va">RNA_feature_subset</span>,</span>
<span>               ADT_feature_subset <span class="op">=</span> <span class="va">ADT_feature_subset</span>,</span>
<span>               cor_method <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>               network_layout <span class="op">=</span> <span class="fu">igraph</span><span class="fu">::</span><span class="va"><a href="https://r.igraph.org/reference/layout_with_fr.html" class="external-link">layout_with_fr</a></span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-31-1.png" width="1152"></p>
<p><strong>Interpretation:</strong></p>
<p>There are roughly 4 groupings in the interaction network,
particularly of interest, CD27 and CD11b is positively correlated with
RPS and RPL ribosomal genes. This could be expected, as activated immune
cells require increased ribosome biogenesis to support the immune
response.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="what-are-the-ligand-receptor-interactions-in-my-data">What are the ligand-receptor interactions in my data?<a class="anchor" aria-label="anchor" href="#what-are-the-ligand-receptor-interactions-in-my-data"></a>
</h3>
<p>Ligand-receptor interactions are the means by which cells communicate
with each other. Ligands (eg. hormones, neurotransmitters, growth
factors etc.) bind to specific receptors on the surface of target cells,
initiating intracellular signalling cascades that alter cell
behaviour.</p>
<p>To take advantage of the ADT expression, CiteFuse implements a
<code>ligandReceptorTest</code> to find ligand-receptor interactions
between sender and receiver cells. Importantly, the ADT count is used to
predict receptor expression within receiver cells. Note that the setting
<code>altExp_name = "RNA"</code> would enable users to predict
ligand-receptor interaction from RNA expression only.</p>
<p>First, we load a database of known ligand-receptor interactions to
query against our data.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the database of known ligand-receptor interactions from cellPhoneDB</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"lr_pair_subset"</span>, package <span class="op">=</span> <span class="st">"CiteFuse"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lr_pair_subset</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1]          [,2]   </span></span>
<span><span class="co">#&gt; [1,] "hg19_IL17RA" "CD45" </span></span>
<span><span class="co">#&gt; [2,] "hg19_FAS"    "CD11b"</span></span>
<span><span class="co">#&gt; [3,] "hg19_GZMK"   "CD62L"</span></span>
<span><span class="co">#&gt; [4,] "hg19_CD40LG" "CD11b"</span></span>
<span><span class="co">#&gt; [5,] "hg19_FLT3LG" "CD62L"</span></span>
<span><span class="co">#&gt; [6,] "hg19_GZMA"   "CD19"</span></span></code></pre></div>
<p>A key challenge in analyzing ligand–receptor interactions is the
difference in scaling and distribution. Comparing unequal distributions
can potentially introduce bias, especially during ligand–receptor
predictions when means are compared. To address this, we first scaled
each feature to a range of 0–1 through min–max normalization and then
transformed the protein expression to force the low-expression values to
zero.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html" class="external-link">normaliseExprs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">sce_citeseq</span>, </span>
<span>                              altExp_name <span class="op">=</span> <span class="st">"ADT"</span>, </span>
<span>                              transform <span class="op">=</span> <span class="st">"zi_minMax"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html" class="external-link">normaliseExprs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">sce_citeseq</span>, </span>
<span>                              altExp_name <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                              exprs_value <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                              transform <span class="op">=</span> <span class="st">"minMax"</span><span class="op">)</span></span></code></pre></div>
<p>For each ligand–receptor interaction pair originating from a cluster
expressing the ligand and another cluster expressing the receptor, a
permutation test is performed on the mean of the average RNA expression
from the ligand cluster and the mean of the protein expression from the
receptor cluster. Only ligand–receptor pairs with a P-value of lower
than 0.05 are defined as significant pairs.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/ligandReceptorTest.html" class="external-link">ligandReceptorTest</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">sce_citeseq</span>,</span>
<span>                                  ligandReceptor_list <span class="op">=</span> <span class="va">lr_pair_subset</span>,</span>
<span>                                  cluster <span class="op">=</span> <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_louvain</span>,</span>
<span>                                  RNA_exprs_value <span class="op">=</span> <span class="st">"minMax"</span>,</span>
<span>                                  use_alt_exp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  altExp_name <span class="op">=</span> <span class="st">"ADT"</span>,</span>
<span>                                  altExp_exprs_value <span class="op">=</span> <span class="st">"zi_minMax"</span>,</span>
<span>                                  num_permute <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span> </span>
<span><span class="co">#&gt; 100 ......200 ......300 ......400 ......500 ......600 ......700 ......800 ......900 ......1000 ......</span></span></code></pre></div>
<p><strong>Alternative visualisation 1:</strong> To visualise the
ligand-receptor predictions, we can plot a heatmap of p-values
indicating significant ligand-receptor pairs between cluster groups.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visLigandReceptor.html" class="external-link">visLigandReceptor</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>                  type <span class="op">=</span> <span class="st">"pval_heatmap"</span>,</span>
<span>                  receptor_type <span class="op">=</span> <span class="st">"ADT"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-35-1.png" width="768"></p>
<p><strong>Alternative visualisation 2:</strong> Here, we plot a network
graph between cluster groups, where the weight of the lines indicate the
number of LR-pairs.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visLigandReceptor.html" class="external-link">visLigandReceptor</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>                  type <span class="op">=</span> <span class="st">"group_network"</span>,</span>
<span>                  receptor_type <span class="op">=</span> <span class="st">"ADT"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-36-1.png" width="768"></p>
<p><strong>Alternative visualisation 3:</strong> Similarly, we can
visualise the above counts of LR-pairs in a heatmap.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visLigandReceptor.html" class="external-link">visLigandReceptor</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>                  type <span class="op">=</span> <span class="st">"group_heatmap"</span>,</span>
<span>                  receptor_type <span class="op">=</span> <span class="st">"ADT"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-37-1.png" width="768"></p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="optional-evaluating-the-importance-of-adt-markers-in-clustering">[Optional] Evaluating the importance of ADT markers in
clustering<a class="anchor" aria-label="anchor" href="#optional-evaluating-the-importance-of-adt-markers-in-clustering"></a>
</h2>
<p>An important evaluation in CITE-seq data analysis is to assess the
quality of each ADT and to evaluate the contribution of ADTs towards
clustering outcome. <code>CiteFuse</code> calculates the relative
importance of ADT towards clustering outcome by using a random forest
model. The higher the score of an ADT, the greater its importance
towards the final clustering outcome.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/importanceADT.html" class="external-link">importanceADT</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>                             group <span class="op">=</span> <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_louvain</span>,</span>
<span>                             subsample <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The importance scores can be visualised in a boxplot and heatmap. Our
evaluation of ADT importance show that unsurprisingly CD4 and CD8 are
the top two discriminating proteins in PBMCs.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visImportance.html" class="external-link">visImportance</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, plot <span class="op">=</span> <span class="st">"boxplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-39-1.png" width="1152"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visImportance.html" class="external-link">visImportance</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, plot <span class="op">=</span> <span class="st">"heatmap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-40-1.png" width="700"></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># view the top 20 markers by decreasing importance</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"importanceADT"</span><span class="op">]</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span></span>
<span><span class="co">#&gt;              CD27               CD8               CD4               CD5 </span></span>
<span><span class="co">#&gt;         40.880877         37.389624         32.358307         12.257423 </span></span>
<span><span class="co">#&gt;              CD28      PECAM (CD31)               CD7             CD11b </span></span>
<span><span class="co">#&gt;         12.054324         11.229291         10.941676         10.603099 </span></span>
<span><span class="co">#&gt; IL7Ralpha (CD127)    MHCII (HLA-DR)              CD44               CD2 </span></span>
<span><span class="co">#&gt;          9.908142          8.508850          8.507701          8.397517 </span></span>
<span><span class="co">#&gt;      CD366 (tim3)         HLA-A,B,C             CD11c               CD3 </span></span>
<span><span class="co">#&gt;          5.697106          5.688288          4.906899          4.769488 </span></span>
<span><span class="co">#&gt;            CD45RA              CD69             CD62L              CD19 </span></span>
<span><span class="co">#&gt;          4.279235          3.791519          3.602994          3.592245</span></span></code></pre></div>
<p>To test whether we would obtain similar clustering results by using a
subset of important markers, we can try integration and clustering with
only ADTs with a score greater than 5.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subset_adt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"importanceADT"</span><span class="op">]</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">subset_adt</span></span>
<span><span class="co">#&gt;  [1] "CD11b"             "CD2"               "CD27"             </span></span>
<span><span class="co">#&gt;  [4] "CD28"              "CD366 (tim3)"      "CD4"              </span></span>
<span><span class="co">#&gt;  [7] "CD44"              "CD5"               "CD7"              </span></span>
<span><span class="co">#&gt; [10] "CD8"               "HLA-A,B,C"         "IL7Ralpha (CD127)"</span></span>
<span><span class="co">#&gt; [13] "MHCII (HLA-DR)"    "PECAM (CD31)"</span></span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/CiteFuse.html" class="external-link">CiteFuse</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                        ADT_subset <span class="op">=</span> <span class="va">subset_adt</span>,</span>
<span>                        metadata_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"W_SNF_adtSubset1"</span>,</span>
<span>                                           <span class="st">"W_ADT_adtSubset1"</span>,</span>
<span>                                           <span class="st">"W_RNA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating affinity matrix </span></span>
<span><span class="co">#&gt; Performing SNF</span></span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="va">SNF_W_clust_adtSubset1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html" class="external-link">spectralClustering</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"W_SNF_adtSubset1"</span><span class="op">]</span><span class="op">]</span>, K <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing Spectral Clustering</span></span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_clust_adtSubset1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">SNF_W_clust_adtSubset1</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span></code></pre></div>
<p>To see how well the integration performed with only a subset of ADT
markers, we can calculate the concordance between the labels (ie.
adjusted rand index, ARI).</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_clust_adtSubset1</span>, <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_clust</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.8499764</span></span></code></pre></div>
<p><strong>Interpretation:</strong></p>
<p>When we compare between the two clustering outcomes, we find that the
ARI is approximately 0.85, where a value of 1 denotes complete
concordance. This high ARI value suggests that the top markers are the
primary contributors to the biological signals in the data,
significantly influencing the integration and clustering patterns.</p>
<p><strong>Discussion:</strong></p>
<ul>
<li>What would the concordance be if we set an even higher threshold for
important ADT markers?</li>
</ul>
<p><br></p>
</div>
<div class="section level2">
<h2 id="optional-performing-multi-sample-analysis-with-citefuse">[Optional] Performing multi-sample analysis with
<code>CiteFuse</code><a class="anchor" aria-label="anchor" href="#optional-performing-multi-sample-analysis-with-citefuse"></a>
</h2>
<p>Lastly, we will jointly analyse the current PBMC CITE-seq data, taken
from healthy human donors, and another subset of CITE-seq data from
patients with cutaneous T-cell lymphoma (CTCL), again from Mimitou et
al. (2019). The data <code>sce_ctcl_subset</code> provided in our
<code>CiteFuse</code> package already contains the clustering
information.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sce_ctcl_subset"</span>, package <span class="op">=</span> <span class="st">"CiteFuse"</span><span class="op">)</span></span></code></pre></div>
<p>To visualise and compare gene or protein expression data, we can use
<code>visualiseExprsList</code> function. Here, we visualise the gene
expression of a subset of genes across the two conditions and
individually for each identified cluster.</p>
<p><strong>Questions:</strong></p>
<ul>
<li>How would we visualise the ADT expression?</li>
</ul>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprsList.html" class="external-link">visualiseExprsList</a></span><span class="op">(</span>sce_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>control <span class="op">=</span> <span class="va">sce_citeseq</span>, ctcl <span class="op">=</span> <span class="va">sce_ctcl_subset</span><span class="op">)</span>,</span>
<span>                   plot <span class="op">=</span> <span class="st">"boxplot"</span>,</span>
<span>                   altExp_name <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>                   exprs_value <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                   feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_S100A10"</span>, <span class="st">"hg19_CD8A"</span><span class="op">)</span>,</span>
<span>                   group_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNF_W_louvain"</span>, <span class="st">"SNF_W_louvain"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-46-1.png" width="700"></p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># fill in code to visualise the ADT expression of CD19 and CD8 across the conditions</span></span></code></pre></div>
<p>We can then perform differential expression analysis of the RNA
expression level across the two clusters that have high CD19 expression
in ADT. Can you gather anything biologically meaningful about the
clusters from looking at the top DE genes?</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEgenesCross.html" class="external-link">DEgenesCross</a></span><span class="op">(</span>sce_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>control <span class="op">=</span> <span class="va">sce_citeseq</span>, ctcl <span class="op">=</span> <span class="va">sce_ctcl_subset</span><span class="op">)</span>,</span>
<span>                       colData_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNF_W_louvain"</span>, <span class="st">"SNF_W_louvain"</span><span class="op">)</span>,</span>
<span>                       group_to_test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2"</span>, <span class="st">"6"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">de_res_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/selectDEgenes.html" class="external-link">selectDEgenes</a></span><span class="op">(</span>de_res <span class="op">=</span> <span class="va">de_res</span><span class="op">)</span></span>
<span><span class="va">de_res_filter</span></span>
<span><span class="co">#&gt; $control</span></span>
<span><span class="co">#&gt;             stats.W         pval     p.adjust meanExprs.1 meanExprs.2</span></span>
<span><span class="co">#&gt; hg19_GNLY      28.0 6.646869e-23 6.646869e-23  0.07220401    4.591530</span></span>
<span><span class="co">#&gt; hg19_CST7     149.5 4.403301e-21 4.403301e-21  0.22512314    2.640981</span></span>
<span><span class="co">#&gt; hg19_NKG7     165.0 8.951222e-21 8.951222e-21  0.51066894    3.815105</span></span>
<span><span class="co">#&gt; hg19_GZMH     301.0 1.984392e-19 1.984392e-19  0.00000000    1.972667</span></span>
<span><span class="co">#&gt; hg19_GZMB     474.0 3.856340e-17 3.856340e-17  0.11667246    2.017016</span></span>
<span><span class="co">#&gt; hg19_CCL5     474.0 1.010639e-16 1.010639e-16  1.04348029    3.772472</span></span>
<span><span class="co">#&gt; hg19_FGFBP2   548.0 1.623043e-16 1.623043e-16  0.03101677    1.666924</span></span>
<span><span class="co">#&gt; hg19_KLRD1    574.5 4.662465e-16 4.662465e-16  0.09327555    1.568567</span></span>
<span><span class="co">#&gt; hg19_EFHD2    794.0 5.121771e-14 5.121771e-14  0.02064028    1.212213</span></span>
<span><span class="co">#&gt; hg19_PRF1     801.5 9.024284e-14 9.024284e-14  0.06473009    1.467354</span></span>
<span><span class="co">#&gt;              meanPct.1 meanPct.2 meanDiff   pctDiff        name   group</span></span>
<span><span class="co">#&gt; hg19_GNLY   0.02325581 0.9927007 4.519326 0.9694449   hg19_GNLY control</span></span>
<span><span class="co">#&gt; hg19_CST7   0.13953488 0.9927007 2.415857 0.8531658   hg19_CST7 control</span></span>
<span><span class="co">#&gt; hg19_NKG7   0.30232558 1.0000000 3.304436 0.6976744   hg19_NKG7 control</span></span>
<span><span class="co">#&gt; hg19_GZMH   0.00000000 0.8978102 1.972667 0.8978102   hg19_GZMH control</span></span>
<span><span class="co">#&gt; hg19_GZMB   0.06976744 0.8832117 1.900344 0.8134442   hg19_GZMB control</span></span>
<span><span class="co">#&gt; hg19_CCL5   0.37209302 1.0000000 2.728992 0.6279070   hg19_CCL5 control</span></span>
<span><span class="co">#&gt; hg19_FGFBP2 0.02325581 0.8248175 1.635907 0.8015617 hg19_FGFBP2 control</span></span>
<span><span class="co">#&gt; hg19_KLRD1  0.09302326 0.8321168 1.475292 0.7390935  hg19_KLRD1 control</span></span>
<span><span class="co">#&gt; hg19_EFHD2  0.02325581 0.7372263 1.191573 0.7139705  hg19_EFHD2 control</span></span>
<span><span class="co">#&gt; hg19_PRF1   0.04651163 0.7591241 1.402624 0.7126125   hg19_PRF1 control</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ctcl</span></span>
<span><span class="co">#&gt;               stats.W         pval     p.adjust meanExprs.1 meanExprs.2</span></span>
<span><span class="co">#&gt; hg19_LTB        368.0 6.626368e-28 6.626368e-28   0.1049424    1.978730</span></span>
<span><span class="co">#&gt; hg19_RPS26        0.0 4.957273e-23 4.957273e-23   1.6292756    5.072471</span></span>
<span><span class="co">#&gt; hg19_IL7R       747.0 5.147332e-21 5.147332e-21   0.1597474    1.651591</span></span>
<span><span class="co">#&gt; hg19_SELL      1006.0 3.198651e-20 3.198651e-20   0.0659935    1.152572</span></span>
<span><span class="co">#&gt; hg19_LEPROTL1   699.0 1.893136e-18 1.893136e-18   0.2381900    1.300745</span></span>
<span><span class="co">#&gt; hg19_EEF1B2     512.5 3.188785e-16 3.188785e-16   1.7119640    3.043623</span></span>
<span><span class="co">#&gt; hg19_HBB       1712.5 1.982149e-15 1.982149e-15   0.0000000    0.431026</span></span>
<span><span class="co">#&gt; hg19_NOSIP     1034.0 2.135217e-14 2.135217e-14   0.2014222    1.157235</span></span>
<span><span class="co">#&gt; hg19_FOS       1278.5 1.229947e-13 1.229947e-13   0.1492607    1.125337</span></span>
<span><span class="co">#&gt; hg19_NPM1       927.0 1.013945e-11 1.013945e-11   1.1724979    2.202268</span></span>
<span><span class="co">#&gt;                meanPct.1 meanPct.2  meanDiff   pctDiff          name group</span></span>
<span><span class="co">#&gt; hg19_LTB      0.08029197 0.9069767 1.8737878 0.8266848      hg19_LTB  ctcl</span></span>
<span><span class="co">#&gt; hg19_RPS26    0.89051095 1.0000000 3.4431949 0.1094891    hg19_RPS26  ctcl</span></span>
<span><span class="co">#&gt; hg19_IL7R     0.10218978 0.8139535 1.4918432 0.7117637     hg19_IL7R  ctcl</span></span>
<span><span class="co">#&gt; hg19_SELL     0.05109489 0.6976744 1.0865780 0.6465795     hg19_SELL  ctcl</span></span>
<span><span class="co">#&gt; hg19_LEPROTL1 0.18978102 0.9069767 1.0625555 0.7171957 hg19_LEPROTL1  ctcl</span></span>
<span><span class="co">#&gt; hg19_EEF1B2   0.86131387 0.9767442 1.3316593 0.1154303   hg19_EEF1B2  ctcl</span></span>
<span><span class="co">#&gt; hg19_HBB      0.00000000 0.4186047 0.4310260 0.4186047      hg19_HBB  ctcl</span></span>
<span><span class="co">#&gt; hg19_NOSIP    0.19708029 0.7674419 0.9558133 0.5703616    hg19_NOSIP  ctcl</span></span>
<span><span class="co">#&gt; hg19_FOS      0.11678832 0.6511628 0.9760767 0.5343745      hg19_FOS  ctcl</span></span>
<span><span class="co">#&gt; hg19_NPM1     0.72992701 0.9534884 1.0297701 0.2235614     hg19_NPM1  ctcl</span></span></code></pre></div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Carissa Chen, Sharon Long, Pengyi Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
