<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SCMultiomeCNRSR">
<title>Analysis of CITE-seq data with CiteFuse • SCMultiomeCNRSR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of CITE-seq data with CiteFuse">
<meta property="og:description" content="SCMultiomeCNRSR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SCMultiomeCNRSR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction_to_single-cell.html">Introduction to single-cell pre-processing</a>
    <a class="dropdown-item" href="../articles/basics_of_single-cell_annotation.html">Basics of single-cell annotation</a>
    <a class="dropdown-item" href="../articles/advancedphenotyping.html">Advanced cell phenotyping</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/carissaynchen/SCMultiome-CNRS-workshop-R">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<link href="citefuse_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="citefuse_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="citefuse_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="citefuse_files/datatables-binding-0.33/datatables.js"></script><link href="citefuse_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="citefuse_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="citefuse_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="citefuse_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="citefuse_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analysis of CITE-seq data with CiteFuse</h1>
                        <h4 data-toc-skip class="author">Carissa Chen<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:cchen@cmri.org.au" class="email"&gt;cchen@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a> Sharon
Long<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:slong@cmri.org.au" class="email"&gt;slong@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>
Pengyi Yang<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:pyang@cmri.org.au" class="email"&gt;pyang@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>3</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/carissaynchen/SCMultiome-CNRS-workshop-R/blob/HEAD/vignettes/citefuse.Rmd" class="external-link"><code>vignettes/citefuse.Rmd</code></a></small>
      <div class="d-none name"><code>citefuse.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>scRNA-seq has allowed us to dissect complex cell populations at the
transcriptomic level, however it lacks additional phenotypic information
such as the quantification of cell-surface proteins which is often
useful to distinguish closely related cell states (eg. in immunology).
CITE-seq uses oligonucleotide-labeled antibodies to integrate cellular
protein and transcriptomic measurements into a single-cell readout. In
this first section where we introduce single-cell multi-modal analysis,
we will learn how to pre-process and integratively explore CITE-seq data
using the <code>CiteFuse</code> package.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="preparation-and-assumed-knowledge">Preparation and assumed knowledge<a class="anchor" aria-label="anchor" href="#preparation-and-assumed-knowledge"></a>
</h3>
<ul>
<li><a href="https://www.nature.com/articles/nmeth.4380" class="external-link">CITE-seq
technology</a></li>
<li><a href="https://doi.org/10.1093/bioinformatics/btaa282" class="external-link">Analysing
CITE-seq with CiteFuse</a></li>
<li><a href="https://sydneybiox.github.io/CiteFuse/" class="external-link">CiteFuse
vignette</a></li>
</ul>
<p><br></p>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>understand the output of CITE-seq data</li>
<li>how to pre-process CITE-seq data</li>
<li>perform exploratory analysis and learn how to jointly visualise
modalities</li>
<li>integrate RNA-ADT information to construct interaction networks</li>
</ul>
<p><br></p>
</div>
<div class="section level3">
<h3 id="schedule">Schedule<a class="anchor" aria-label="anchor" href="#schedule"></a>
</h3>
<p>Structure of this section:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction to <code>CiteFuse</code>
</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Integration and clustering of modalities</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Exploratory analysis and differential gene detection</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Integrative analysis of RNA and ADT modalities</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Concluding remarks</td>
<td></td>
</tr>
</tbody>
</table>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The CITE-seq protocol involves conjugating antibodies to
oligonucleotides, where each antibody is specific to a cell surface
protein marker. The antibodies bind to their specific protein marker,
and mRNA is lysed from the cell. Both molecular species are captured for
scRNA-sequencing. In a process called ‘cell hashing’, cells are stained
with sample-specific oligonucleotide-labelled antibodies (‘hashing
antibodies’), thus allowing the samples to be multiplexed and sequenced
in a single experiment and later traced back to their origin. Thus,
allowing for high-throughput and high-dimensional multi-omic single-cell
analysis.</p>
<p><code>CiteFuse</code> is a computational framework that implements a
suite of methods and tools for CITE-seq data from pre-processing through
to integrative analytics. This includes doublet detection, network-based
modality integration, cell type clustering, differential RNA and protein
expression analysis, ADT evaluation, ligand-receptor interaction
analysis, and interactive web-based visualisation of the analyses.</p>
<p>In this workshop, we demonstrate the usage of <code>CiteFuse</code>
on a subset data of CITE-seq data from human PBMCs as an example <a href="https://doi.org/10.1038/s41592-019-0392-0" class="external-link">Mimitou et al.,
2019</a>.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="loading-libraries-and-the-data">Loading libraries and the data<a class="anchor" aria-label="anchor" href="#loading-libraries-and-the-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CiteFuse</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/DT" class="external-link">DT</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mclust-org.github.io/mclust/" class="external-link">mclust</a></span><span class="op">)</span></span></code></pre></div>
<p>Here, we start from a list of three matrices of unique molecular
identifier (UMI), antibody derived tags (ADT) and hashtag
oligonucleotide (HTO) count, which have common cell names.</p>
<p>There are 500 cells in our subsetted dataset. And characteristically
of CITE-seq data, the matrices are matched, meaning that for any given
cell we know the expression level of their RNA transcripts (genome-wide)
and its corresponding cell surface protein expression. The
<code>preprocessing</code> function will utilise the three matrices and
its common cell names to create a <code>SingleCellExperiment</code>
object, which stores RNA data in an <code>assay</code> and
<code>ADT</code> and <code>HTO</code> data within in the
<code>altExp</code> slot.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"CITEseq_example"</span>, package <span class="op">=</span> <span class="st">"CiteFuse"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">CITEseq_example</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RNA" "ADT" "HTO"</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">CITEseq_example</span>, <span class="va">dim</span><span class="op">)</span></span>
<span><span class="co">#&gt; $RNA</span></span>
<span><span class="co">#&gt; [1] 19521   500</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ADT</span></span>
<span><span class="co">#&gt; [1]  49 500</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $HTO</span></span>
<span><span class="co">#&gt; [1]   4 500</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/preprocessing.html" class="external-link">preprocessing</a></span><span class="op">(</span><span class="va">CITEseq_example</span><span class="op">)</span></span>
<span><span class="va">sce_citeseq</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 19521 500 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(19521): hg19_AL627309.1 hg19_AL669831.5 ... hg19_MT-ND6</span></span>
<span><span class="co">#&gt;   hg19_MT-CYB</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(500): AAGCCGCGTTGTCTTT GATCGCGGTTATCGGT ... TTGGCAACACTAGTAC</span></span>
<span><span class="co">#&gt;   GCTGCGAGTTGTGGCC</span></span>
<span><span class="co">#&gt; colData names(0):</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(2): ADT HTO</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="integration-of-rna-and-adt-using-snf">Integration of RNA and ADT using <code>SNF</code><a class="anchor" aria-label="anchor" href="#integration-of-rna-and-adt-using-snf"></a>
</h2>
<p>The next step of analysis is to integrate the RNA and ADT matrix. We
use a popular integration algorithm called similarity network fusion
(SNF) to integrate the multi-omic data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html" class="external-link">normaliseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, altExp_name <span class="op">=</span> <span class="st">"ADT"</span>, transform <span class="op">=</span> <span class="st">"log"</span><span class="op">)</span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/CiteFuse.html" class="external-link">CiteFuse</a></span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating affinity matrix </span></span>
<span><span class="co">#&gt; Performing SNF</span></span></code></pre></div>
<p>We now proceed with the fused matrix, which is stored as
<code>SNF_W</code> in the metadata slot of our <code>sce_citeseq</code>
object. We can calculate the t-SNE to generate an integrated embedding
using the SNF network.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/reducedDimSNF.html" class="external-link">reducedDimSNF</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                             method <span class="op">=</span> <span class="st">"tSNE"</span>, </span>
<span>                             dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="how-many-clusters-are-in-my-data">How many clusters are in my data?<a class="anchor" aria-label="anchor" href="#how-many-clusters-are-in-my-data"></a>
</h2>
<div class="section level3">
<h3 id="clustering-using-louvain-clustering">Clustering using Louvain clustering<a class="anchor" aria-label="anchor" href="#clustering-using-louvain-clustering"></a>
</h3>
<p>As well as spectral clustering, CiteFuse can implement Louvain
clustering if users wish to use another clustering method. We use the
<code>igraph</code> package, and any community detection algorithms
available in their package can be selected by changing the
<code>method</code> parameter.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="va">SNF_W_louvain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/igraphClustering.html" class="external-link">igraphClustering</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, method <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">SNF_W_louvain</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNF_W_louvain</span></span>
<span><span class="co">#&gt;   1   2   3   4   5   6   7   8 </span></span>
<span><span class="co">#&gt;  37  96 146  64  45  34  58  20</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_louvain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">SNF_W_louvain</span><span class="op">)</span></span></code></pre></div>
<p>The outcome of the clustering can be easily visualised on a reduced
dimensions plot by highlighting the points by cluster label.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span>, colour_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"tSNE (SNF louvain clustering)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-7-1.png" width="1152"></p>
<p>Alternatively, the clustering can be easily visualised using a KNN
graph labelled by cluster number.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseKNN.html" class="external-link">visualiseKNN</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, colour_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="optional-clustering-using-spectral-clustering">[Optional] Clustering using Spectral clustering<a class="anchor" aria-label="anchor" href="#optional-clustering-using-spectral-clustering"></a>
</h3>
<p>CiteFuse implements two different clustering algorithms on the fused
matrix, spectral clustering and Louvain clustering. First, we perform
spectral clustering with sufficient numbers of <code>K</code> and use
the eigen values to determine the optimal number of clusters. What is
the optimal number of clusters here?</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SNF_W_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html" class="external-link">spectralClustering</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"SNF_W"</span><span class="op">]</span><span class="op">]</span>, K <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing Spectral Clustering</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">SNF_W_clust</span><span class="op">$</span><span class="va">eigen_values</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">SNF_W_clust</span><span class="op">$</span><span class="va">eigen_values</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 5</span></span></code></pre></div>
<p>Using the optimal cluster number defined from the previous step, we
can now use the <code>spectralClustering</code> function to cluster the
single cells by specifying the number of clusters in <code>K</code>. The
function takes a cell-to-cell similarity matrix as an input. We have
already created the fused similarity matrix from <code>CiteFuse</code>.
Since the <code>CiteFuse</code> function creates and stores the
similarity matries from ADT and RNA expression, as well the fused
matrix, we can use these two to compare the clustering outcomes by data
modality.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SNF_W_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html" class="external-link">spectralClustering</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"SNF_W"</span><span class="op">]</span><span class="op">]</span>, K <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing Spectral Clustering</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">SNF_W_clust</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SNF_W1_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html" class="external-link">spectralClustering</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"ADT_W"</span><span class="op">]</span><span class="op">]</span>, K <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing Spectral Clustering</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">ADT_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">SNF_W1_clust</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SNF_W2_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/spectralClustering.html" class="external-link">spectralClustering</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"RNA_W"</span><span class="op">]</span><span class="op">]</span>, K <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing Spectral Clustering</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span><span class="op">$</span><span class="va">RNA_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">SNF_W2_clust</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span>, colour_by <span class="op">=</span> <span class="st">"SNF_W_clust"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"tSNE (SNF spectral clustering)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span>, colour_by <span class="op">=</span> <span class="st">"ADT_clust"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"tSNE (ADT spectral clustering)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span>, colour_by <span class="op">=</span> <span class="st">"RNA_clust"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"tSNE (RNA spectral clustering)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g1</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">g2</span> <span class="op">+</span> <span class="va">g3</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-11-2.png" width="700"></p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="exploratory-analysis-of-feature-expression-across-modalities">Exploratory analysis of feature expression across modalities<a class="anchor" aria-label="anchor" href="#exploratory-analysis-of-feature-expression-across-modalities"></a>
</h2>
<p>CiteFuse has a wide range of visualisation tools to facilitate
exploratory analysis of CITE-seq data. The <code>visualiseExprs</code>
function is an easy-to-use function to generate boxplots, violin plots,
jitter plots, density plots, and pairwise scatter/density plots of genes
and proteins expressed in the data. The plots can be grouped by using
the cluster labels stored in the <code>sce_citeseq</code> object.</p>
<p>The expression of genes and proteins can be visualised by changing
the <code>colour_by</code> parameter to assess the clusters. As an
example, we highlight the plot by the RNA and ADT expression level of
CD8.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span>, </span>
<span>                   colour_by <span class="op">=</span> <span class="st">"hg19_CD8A"</span>,</span>
<span>                   data_from <span class="op">=</span> <span class="st">"assay"</span>,</span>
<span>                   assay_name <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"tSNE: hg19_CD8A (RNA expression)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseDim.html" class="external-link">visualiseDim</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,dimNames <span class="op">=</span> <span class="st">"tSNE_joint"</span>, </span>
<span>                   colour_by <span class="op">=</span> <span class="st">"CD8"</span>,</span>
<span>                   data_from <span class="op">=</span> <span class="st">"altExp"</span>,</span>
<span>                   altExp_assay_name <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"tSNE: CD8 (ADT expression)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g1</span> <span class="op">+</span> <span class="va">g2</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-12-1.png" width="1152"></p>
<p><strong>Alternative visualisation 1:</strong> We can plot a boxplot
of the expression of a subset of genes per cluster detected by the
louvain clustering algorithm.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html" class="external-link">visualiseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>               plot <span class="op">=</span> <span class="st">"boxplot"</span>, </span>
<span>               altExp_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>               group_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span>,</span>
<span>               feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-13-1.png" width="768"></p>
<p><strong>Alternative visualisation 2:</strong> We can plot a violin
plot of the expression of a subset of genes per cluster detected by the
louvain clustering algorithm.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html" class="external-link">visualiseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>               plot <span class="op">=</span> <span class="st">"violin"</span>,</span>
<span>               altExp_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>               group_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span>,</span>
<span>               feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p><strong>Alternative visualisation 3:</strong> We can plot a jitter of
the expression of a subset of genes per cluster detected by the louvain
clustering algorithm.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html" class="external-link">visualiseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>               plot <span class="op">=</span> <span class="st">"jitter"</span>, </span>
<span>               altExp_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>               group_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span>,</span>
<span>               feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p><strong>Alternative visualisation 4:</strong> We can plot a density
plot of the expression of a subset of genes per cluster detected by the
louvain clustering algorithm.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html" class="external-link">visualiseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>               plot <span class="op">=</span> <span class="st">"density"</span>, </span>
<span>               altExp_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>               group_by <span class="op">=</span> <span class="st">"SNF_W_louvain"</span>,</span>
<span>               feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_CD2"</span>, <span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_CD19"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p><strong>Questions:</strong></p>
<ul>
<li>From these collection of visualisations, do you notice any markers
that are differentially expressed in any clusters?</li>
<li>How would you plot the above, but for ADT?</li>
</ul>
<p><strong>Alternative visualisation 5:</strong> We can also create a
pairwise scatter plot to visualise the distribution of cells that are
expressing certain markers.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visualiseExprs.html" class="external-link">visualiseExprs</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, altExp_name <span class="op">=</span> <span class="st">"ADT"</span>, </span>
<span>               plot <span class="op">=</span> <span class="st">"pairwise"</span>, </span>
<span>               feature_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; number of iterations= 14 </span></span>
<span><span class="co">#&gt; number of iterations= 22</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<p><strong>Questions:</strong></p>
<ul>
<li>What do each of the quadrants correspond to?</li>
<li>Do we have many cells that co-express both markers?</li>
</ul>
<p><br></p>
</div>
<div class="section level2">
<h2 id="what-genes-are-differentially-expressed-between-my-clusters">What genes are differentially expressed between my clusters?<a class="anchor" aria-label="anchor" href="#what-genes-are-differentially-expressed-between-my-clusters"></a>
</h2>
<div class="section level3">
<h3 id="perform-de-analysis-with-wilcoxon-rank-sum-test">Perform DE Analysis with Wilcoxon Rank Sum test<a class="anchor" aria-label="anchor" href="#perform-de-analysis-with-wilcoxon-rank-sum-test"></a>
</h3>
<p><code>CiteFuse</code> can also calculates differentially expressed
(DE) genes through the <code>DEgenes</code> function. The cluster
grouping to use must be specified in the <code>group</code> parameter.
If <code>altExp_name</code> is not specified, RNA expression will be
used as the default expression matrix.</p>
<p>Results form the DE analysis is stored in <code>sce_citeseq</code> as
<code>DE_res_RNA_filter</code> and <code>DE_res_ADT_filter</code> for
RNA and ADT expression, respectively.</p>
<div class="section level4">
<h4 id="for-rna-modality">For RNA modality<a class="anchor" aria-label="anchor" href="#for-rna-modality"></a>
</h4>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DE will be performed for RNA if altExp_name = "none" </span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEgenes.html" class="external-link">DEgenes</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                       altExp_name <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                       group <span class="op">=</span> <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_louvain</span>,</span>
<span>                       return_all <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       exprs_pct <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/selectDEgenes.html" class="external-link">selectDEgenes</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                             altExp_name <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"DE_res_RNA_filter"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                 digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["1.hg19_FCN1","1.hg19_LYZ","1.hg19_S100A9","1.hg19_CLEC7A","1.hg19_CSTA","1.hg19_S100A8","1.hg19_SERPINA1","1.hg19_MNDA","1.hg19_VCAN","1.hg19_SPI1","2.hg19_LTB","2.hg19_TCF7","2.hg19_IL7R","2.hg19_LDHB","2.hg19_CD27","2.hg19_LEF1","2.hg19_CAMK4","2.hg19_SELL","2.hg19_NPM1","2.hg19_NOSIP","3.hg19_CD8A","3.hg19_CCL5","3.hg19_GZMH","3.hg19_GNLY","3.hg19_KLRD1","3.hg19_NKG7","3.hg19_CST7","3.hg19_FGFBP2","3.hg19_GZMB","3.hg19_PRF1","4.hg19_DUSP2","4.hg19_KLRB1","4.hg19_KLRG1","4.hg19_LYAR","4.hg19_IL7R","4.hg19_CMC1","4.hg19_GZMM","4.hg19_IL32","4.hg19_HCST","4.hg19_CNN2","5.hg19_ACTN1","5.hg19_CCR7","5.hg19_LEF1","5.hg19_CD8B","5.hg19_TCF7","5.hg19_PIK3IP1","5.hg19_RCAN3","5.hg19_LDLRAP1","5.hg19_LTB","5.hg19_LDHB","6.hg19_IL7R","6.hg19_S100A10","6.hg19_IL32","6.hg19_RORA","6.hg19_FAM107B","6.hg19_GUK1","6.hg19_CD2","6.hg19_CD6","6.hg19_ARHGAP15","6.hg19_KLRB1","7.hg19_SH2D1B","7.hg19_KLRF1","7.hg19_CLIC3","7.hg19_TRDC","7.hg19_MYOM2","7.hg19_SPON2","7.hg19_FCER1G","7.hg19_TYROBP","7.hg19_IL2RB","7.hg19_KLRB1","8.hg19_BANK1","8.hg19_BLNK","8.hg19_POU2AF1","8.hg19_CD79A","8.hg19_MS4A1","8.hg19_LINC00926","8.hg19_BLK","8.hg19_CD180","8.hg19_JCHAIN","8.hg19_MZB1"],["  465","  465","  931"," 1406"," 1620"," 1407"," 1630"," 2315"," 2315","  967"," 5390"," 8006"," 7300"," 7925","11423","11132","11730","11078"," 9590","10806"," 7396"," 6265"," 7982"," 8726","10119"," 9049"," 9417","10972","10658","13761"," 6806"," 8963"," 8687"," 8966"," 8899"," 9903"," 9611"," 9237"," 9706","10070"," 5514"," 5743"," 4312"," 4739"," 4874"," 5376"," 6006"," 6280"," 5131"," 4889"," 3256"," 4469"," 4659"," 5271"," 5324"," 5188"," 5316"," 5590"," 5528"," 5922"," 4378"," 3455"," 4267"," 5195"," 2670"," 3468"," 4928"," 3909"," 4442"," 4451"," 1215"," 1680"," 1920","  748"," 1452"," 2168"," 1958"," 2169"," 2175"," 2182"],["2.4e-101","2.4e-101"," 4.8e-95"," 1.4e-88"," 1.5e-88"," 8.5e-84"," 3.0e-83"," 2.1e-79"," 2.1e-79"," 7.8e-78"," 5.0e-34"," 2.5e-28"," 5.1e-25"," 1.2e-20"," 1.6e-19"," 4.5e-18"," 2.4e-15"," 7.9e-15"," 9.8e-15"," 2.1e-14"," 7.4e-43"," 2.4e-42"," 4.8e-40"," 3.7e-33"," 9.7e-32"," 1.3e-31"," 1.5e-30"," 4.0e-29"," 5.7e-29"," 3.9e-19"," 4.3e-14"," 6.3e-09"," 9.1e-09"," 1.5e-07"," 3.6e-07"," 3.7e-06"," 1.1e-05"," 1.1e-05"," 7.6e-05"," 9.1e-05"," 1.4e-19"," 2.3e-18"," 1.2e-17"," 4.4e-14"," 8.3e-13"," 1.4e-12"," 1.0e-11"," 3.2e-11"," 1.0e-09"," 2.2e-09"," 4.5e-10"," 2.0e-05"," 5.4e-05"," 1.4e-04"," 3.2e-04"," 5.4e-04"," 6.0e-04"," 8.1e-04"," 1.2e-03"," 2.0e-03"," 2.1e-58"," 6.7e-40"," 6.6e-37"," 1.9e-36"," 2.6e-36"," 6.4e-31"," 2.7e-30"," 1.1e-26"," 1.2e-25"," 2.8e-24"," 1.1e-72"," 1.9e-71"," 5.7e-66"," 2.1e-64"," 2.5e-60"," 2.2e-55"," 4.2e-52"," 2.4e-51"," 4.1e-51"," 2.0e-47"],["2.4e-101","2.4e-101"," 4.8e-95"," 1.4e-88"," 1.5e-88"," 8.5e-84"," 3.0e-83"," 2.1e-79"," 2.1e-79"," 7.8e-78"," 5.0e-34"," 2.5e-28"," 5.1e-25"," 1.2e-20"," 1.6e-19"," 4.5e-18"," 2.4e-15"," 7.9e-15"," 9.8e-15"," 2.1e-14"," 7.4e-43"," 2.4e-42"," 4.8e-40"," 3.7e-33"," 9.7e-32"," 1.3e-31"," 1.5e-30"," 4.0e-29"," 5.7e-29"," 3.9e-19"," 4.3e-14"," 6.3e-09"," 9.1e-09"," 1.5e-07"," 3.6e-07"," 3.7e-06"," 1.1e-05"," 1.1e-05"," 7.6e-05"," 9.1e-05"," 1.4e-19"," 2.3e-18"," 1.2e-17"," 4.4e-14"," 8.3e-13"," 1.4e-12"," 1.0e-11"," 3.2e-11"," 1.0e-09"," 2.2e-09"," 4.5e-10"," 2.0e-05"," 5.4e-05"," 1.4e-04"," 3.2e-04"," 5.4e-04"," 6.0e-04"," 8.1e-04"," 1.2e-03"," 2.0e-03"," 2.1e-58"," 6.7e-40"," 6.6e-37"," 1.9e-36"," 2.6e-36"," 6.4e-31"," 2.7e-30"," 1.1e-26"," 1.2e-25"," 2.8e-24"," 1.1e-72"," 1.9e-71"," 5.7e-66"," 2.1e-64"," 2.5e-60"," 2.2e-55"," 4.2e-52"," 2.4e-51"," 4.1e-51"," 2.0e-47"],["0.0022","0.0015","0.0019","0.0028","0.0000","0.0074","0.0028","0.0000","0.0000","0.0167","0.6549","0.2711","0.8047","0.8165","0.1398","0.2100","0.1979","0.3799","1.3506","0.4141","0.4778","1.4971","0.5373","1.8616","0.5119","1.6816","1.1605","0.5340","0.7239","0.6331","0.4860","0.5076","0.4693","0.4857","0.9687","0.4535","0.6025","1.7851","1.3946","0.5800","0.0795","0.0900","0.2586","0.3803","0.3804","0.2251","0.1689","0.1451","0.9043","0.9382","0.9924","1.3815","1.8219","0.4506","0.4818","0.8269","0.6790","0.4406","0.5450","0.6150","0.0200","0.1488","0.1034","0.0988","0.2706","0.2960","0.2561","0.5486","0.2426","0.4621","0.0048","0.0000","0.0000","0.0197","0.0081","0.0030","0.0063","0.0019","0.0051","0.0068"],["2.50","3.85","3.37","1.15","1.43","3.67","1.50","1.18","2.03","1.61","2.47","1.21","2.29","1.88","0.65","0.84","0.69","1.10","2.16","1.07","1.99","3.93","2.08","4.74","1.68","3.98","2.77","1.74","2.12","1.55","1.43","1.58","1.11","1.06","1.92","1.03","1.08","2.56","1.93","0.99","0.58","0.63","1.06","1.38","1.17","0.79","0.58","0.57","2.00","1.86","2.44","2.09","2.73","0.94","0.90","1.26","1.15","0.81","0.95","1.06","1.00","1.34","1.52","1.06","1.93","1.86","1.77","2.16","1.37","2.04","1.13","0.51","0.62","1.75","2.01","0.98","0.52","0.56","1.43","1.09"],["0.0022","0.0022","0.0022","0.0022","0.0000","0.0065","0.0043","0.0000","0.0000","0.0194","0.3168","0.1931","0.3490","0.5371","0.1188","0.1609","0.1757","0.2599","0.7426","0.3267","0.2768","0.4972","0.2768","0.4605","0.2881","0.4972","0.4718","0.2740","0.3079","0.3390","0.3303","0.2477","0.3005","0.3509","0.4312","0.2661","0.4128","0.7317","0.7569","0.4358","0.0791","0.0769","0.1890","0.2308","0.2527","0.1890","0.1429","0.1297","0.3846","0.5692","0.4292","0.7597","0.7468","0.3369","0.3863","0.6030","0.4785","0.3369","0.4270","0.2639","0.0181","0.1086","0.0973","0.0611","0.1878","0.1968","0.1063","0.2081","0.1968","0.2127","0.0042","0.0000","0.0000","0.0167","0.0083","0.0021","0.0063","0.0042","0.0042","0.0063"],["0.95","0.95","0.89","0.84","0.81","0.84","0.81","0.73","0.73","0.89","0.93","0.75","0.95","0.85","0.52","0.57","0.56","0.66","0.96","0.69","0.88","1.00","0.90","0.99","0.84","1.00","0.99","0.82","0.88","0.77","0.77","0.53","0.66","0.66","0.69","0.53","0.70","0.95","0.91","0.64","0.53","0.51","0.76","0.73","0.78","0.69","0.58","0.51","0.93","0.89","0.94","0.97","0.94","0.62","0.65","0.82","0.74","0.62","0.71","0.56","0.67","0.81","0.71","0.66","0.88","0.83","0.72","0.95","0.76","0.83","0.75","0.65","0.60","0.85","0.70","0.55","0.60","0.55","0.55","0.55"],["2.49","3.84","3.37","1.15","1.43","3.66","1.49","1.18","2.03","1.59","1.81","0.94","1.49","1.06","0.52","0.63","0.49","0.72","0.81","0.66","1.51","2.44","1.54","2.87","1.17","2.30","1.61","1.20","1.40","0.92","0.94","1.08","0.64","0.57","0.95","0.57","0.48","0.77","0.54","0.41","0.50","0.54","0.80","1.00","0.79","0.56","0.41","0.42","1.10","0.92","1.44","0.71","0.91","0.49","0.42","0.44","0.47","0.37","0.41","0.45","0.98","1.20","1.42","0.96","1.66","1.57","1.52","1.61","1.12","1.58","1.12","0.51","0.62","1.73","2.00","0.98","0.51","0.56","1.43","1.08"],["0.94","0.94","0.89","0.84","0.81","0.83","0.81","0.73","0.73","0.87","0.61","0.56","0.60","0.32","0.40","0.41","0.39","0.40","0.22","0.36","0.61","0.50","0.63","0.53","0.55","0.50","0.52","0.54","0.58","0.43","0.44","0.28","0.36","0.31","0.26","0.27","0.29","0.22","0.15","0.20","0.45","0.43","0.57","0.50","0.53","0.50","0.43","0.38","0.55","0.32","0.51","0.21","0.19","0.28","0.26","0.22","0.26","0.28","0.28","0.29","0.65","0.70","0.61","0.59","0.69","0.63","0.62","0.74","0.56","0.61","0.75","0.65","0.60","0.83","0.69","0.55","0.59","0.55","0.55","0.54"],["hg19_FCN1","hg19_LYZ","hg19_S100A9","hg19_CLEC7A","hg19_CSTA","hg19_S100A8","hg19_SERPINA1","hg19_MNDA","hg19_VCAN","hg19_SPI1","hg19_LTB","hg19_TCF7","hg19_IL7R","hg19_LDHB","hg19_CD27","hg19_LEF1","hg19_CAMK4","hg19_SELL","hg19_NPM1","hg19_NOSIP","hg19_CD8A","hg19_CCL5","hg19_GZMH","hg19_GNLY","hg19_KLRD1","hg19_NKG7","hg19_CST7","hg19_FGFBP2","hg19_GZMB","hg19_PRF1","hg19_DUSP2","hg19_KLRB1","hg19_KLRG1","hg19_LYAR","hg19_IL7R","hg19_CMC1","hg19_GZMM","hg19_IL32","hg19_HCST","hg19_CNN2","hg19_ACTN1","hg19_CCR7","hg19_LEF1","hg19_CD8B","hg19_TCF7","hg19_PIK3IP1","hg19_RCAN3","hg19_LDLRAP1","hg19_LTB","hg19_LDHB","hg19_IL7R","hg19_S100A10","hg19_IL32","hg19_RORA","hg19_FAM107B","hg19_GUK1","hg19_CD2","hg19_CD6","hg19_ARHGAP15","hg19_KLRB1","hg19_SH2D1B","hg19_KLRF1","hg19_CLIC3","hg19_TRDC","hg19_MYOM2","hg19_SPON2","hg19_FCER1G","hg19_TYROBP","hg19_IL2RB","hg19_KLRB1","hg19_BANK1","hg19_BLNK","hg19_POU2AF1","hg19_CD79A","hg19_MS4A1","hg19_LINC00926","hg19_BLK","hg19_CD180","hg19_JCHAIN","hg19_MZB1"],["1","1","1","1","1","1","1","1","1","1","2","2","2","2","2","2","2","2","2","2","3","3","3","3","3","3","3","3","3","3","4","4","4","4","4","4","4","4","4","4","5","5","5","5","5","5","5","5","5","5","6","6","6","6","6","6","6","6","6","6","7","7","7","7","7","7","7","7","7","7","8","8","8","8","8","8","8","8","8","8"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stats.W<\/th>\n      <th>pval<\/th>\n      <th>p.adjust<\/th>\n      <th>meanExprs.1<\/th>\n      <th>meanExprs.2<\/th>\n      <th>meanPct.1<\/th>\n      <th>meanPct.2<\/th>\n      <th>meanDiff<\/th>\n      <th>pctDiff<\/th>\n      <th>name<\/th>\n      <th>group<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"stats.W","targets":1},{"name":"pval","targets":2},{"name":"p.adjust","targets":3},{"name":"meanExprs.1","targets":4},{"name":"meanExprs.2","targets":5},{"name":"meanPct.1","targets":6},{"name":"meanPct.2","targets":7},{"name":"meanDiff","targets":8},{"name":"pctDiff","targets":9},{"name":"name","targets":10},{"name":"group","targets":11}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level4">
<h4 id="for-adt-modality">For ADT modality<a class="anchor" aria-label="anchor" href="#for-adt-modality"></a>
</h4>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEgenes.html" class="external-link">DEgenes</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                       altExp_name <span class="op">=</span> <span class="st">"ADT"</span>, </span>
<span>                       group <span class="op">=</span> <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_louvain</span>,</span>
<span>                       return_all <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       exprs_pct <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/selectDEgenes.html" class="external-link">selectDEgenes</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>                             altExp_name <span class="op">=</span> <span class="st">"ADT"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"DE_res_ADT_filter"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                 digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"filter":"none","vertical":false,"data":[["1.CD366 (tim3)","1.LAMP1","1.CD14","1.CD45","1.B7-H1 (PD-L1)","1.TCRg","1.CD45RO","1.CD1a","1.B220 (CD45R)","1.EpCAM (CD326)","2.CD28","2.CD26 (Adenosine)","2.CD3","4","5","7","8.CD45RO","8.Ox40 (CD134)","8.CD1a","8.LAMP1"],[" 1020"," 1174"," 2084"," 1874"," 2374"," 2263"," 2106"," 2211"," 2454"," 2904"," 5874","13724","13502","10418"," 6716"," 4840"," 2496"," 2943"," 3085"," 3394"],["1.8e-20","7.9e-19","1.4e-18","1.8e-16","2.2e-15","7.4e-15","1.2e-14","2.6e-14","1.3e-13","2.3e-12","1.4e-26","9.6e-07","3.4e-06","3.2e-04","1.3e-04","1.1e-15","2.4e-04","1.6e-03","6.1e-03","2.4e-02"],["1.8e-20","7.9e-19","1.4e-18","1.8e-16","2.2e-15","7.4e-15","1.2e-14","2.6e-14","1.3e-13","2.3e-12","1.4e-26","9.6e-07","3.4e-06","3.2e-04","1.3e-04","1.1e-15","2.4e-04","1.6e-03","6.1e-03","2.4e-02"],["0.68","0.98","0.31","0.61","0.44","0.60","1.22","1.13","0.84","0.58","1.36","0.49","1.44","0.50","1.54","0.73","1.30","0.51","1.21","1.11"],["2.84","2.82","1.99","1.61","1.50","1.63","2.67","2.32","1.77","1.58","2.57","0.89","1.84","1.02","2.12","1.69","2.05","0.93","1.58","1.44"],["0.58","0.79","0.35","0.60","0.46","0.58","0.87","0.89","0.76","0.58","0.83","0.40","0.85","0.43","0.85","0.56","0.88","0.51","0.89","0.80"],["0.97","1.00","0.86","0.97","0.92","1.00","1.00","1.00","1.00","0.89","0.99","0.65","0.97","0.59","1.00","0.95","1.00","0.85","1.00","1.00"],["2.16","1.84","1.67","1.00","1.05","1.03","1.45","1.19","0.93","1.00","1.21","0.40","0.41","0.51","0.57","0.96","0.75","0.41","0.38","0.33"],["0.40","0.21","0.51","0.37","0.45","0.42","0.13","0.11","0.24","0.31","0.16","0.24","0.12","0.16","0.15","0.39","0.12","0.34","0.11","0.20"],["CD366 (tim3)","LAMP1","CD14","CD45","B7-H1 (PD-L1)","TCRg","CD45RO","CD1a","B220 (CD45R)","EpCAM (CD326)","CD28","CD26 (Adenosine)","CD3","CD26 (Adenosine)","CD28","CD366 (tim3)","CD45RO","Ox40 (CD134)","CD1a","LAMP1"],["1","1","1","1","1","1","1","1","1","1","2","2","2","4","5","7","8","8","8","8"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stats.W<\/th>\n      <th>pval<\/th>\n      <th>p.adjust<\/th>\n      <th>meanExprs.1<\/th>\n      <th>meanExprs.2<\/th>\n      <th>meanPct.1<\/th>\n      <th>meanPct.2<\/th>\n      <th>meanDiff<\/th>\n      <th>pctDiff<\/th>\n      <th>name<\/th>\n      <th>group<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"stats.W","targets":1},{"name":"pval","targets":2},{"name":"p.adjust","targets":3},{"name":"meanExprs.1","targets":4},{"name":"meanExprs.2","targets":5},{"name":"meanPct.1","targets":6},{"name":"meanPct.2","targets":7},{"name":"meanDiff","targets":8},{"name":"pctDiff","targets":9},{"name":"name","targets":10},{"name":"group","targets":11}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="visualising-de-results">Visualising DE Results<a class="anchor" aria-label="anchor" href="#visualising-de-results"></a>
</h3>
<p>After running DE analysis, we obtain the p-value and relevant
statistics to rank our differentially expressed genes in each cluster.
To visualise a selection of top DE genes, we can use
`<code>DEcomparisonPlot</code>. This lollipop plot shows the -log10
transformed adjusted p-values of the specified genes in RNA and ADT in
each cluster.</p>
<p>It requires a list of features from both RNA and ADT that we are
interested in, as well as the output from the DE analysis.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19_CD4"</span>, <span class="st">"hg19_CD8A"</span>, <span class="st">"hg19_HLA-DRB1"</span>, <span class="st">"hg19_ITGAX"</span>, <span class="st">"hg19_NCAM1"</span>, <span class="st">"hg19_CD27"</span>, <span class="st">"hg19_CD19"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">adt_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"MHCII (HLA-DR)"</span>, <span class="st">"CD11c"</span>, <span class="st">"CD56"</span>, <span class="st">"CD27"</span>, <span class="st">"CD19"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rna_DEgenes_all</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"DE_res_RNA"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">adt_DEgenes_all</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"DE_res_ADT"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">feature_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>RNA <span class="op">=</span> <span class="va">rna_list</span>, ADT <span class="op">=</span> <span class="va">adt_list</span><span class="op">)</span></span>
<span><span class="va">de_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>RNA <span class="op">=</span> <span class="va">rna_DEgenes_all</span>, ADT <span class="op">=</span> <span class="va">adt_DEgenes_all</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/DEcomparisonPlot.html" class="external-link">DEcomparisonPlot</a></span><span class="op">(</span>de_list <span class="op">=</span> <span class="va">de_list</span>,</span>
<span>                 feature_list <span class="op">=</span> <span class="va">feature_list</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-20-1.png" width="768"></p>
<p><strong>Interpretation:</strong></p>
<p>For each louvain-generated groups, we compare the -log10 transformed
p-values of RNA genes (right) and ADT markers (left). The larger the
value, the more significant the p-value. From this, we can see that in
cluster 7, there is the significant differential expression of MHCII and
CD19. This suggests that cluster 7 are B cells, possibly a B cell
subtype.</p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="how-do-i-analyse-the-rna-and-adt-modalities-together">How do I analyse the RNA and ADT modalities together?<a class="anchor" aria-label="anchor" href="#how-do-i-analyse-the-rna-and-adt-modalities-together"></a>
</h2>
<div class="section level3">
<h3 id="is-there-a-relationship-between-important-molecular-factors-in-rna-and-adt">Is there a relationship between important molecular factors in RNA
and ADT?<a class="anchor" aria-label="anchor" href="#is-there-a-relationship-between-important-molecular-factors-in-rna-and-adt"></a>
</h3>
<p>We can generate a network diagram using the
<code>geneADTnetwork</code> function to examine the association between
the differentially expressed genes and protein markers. We specify the
unique DE genes across all clusters from the RNA and ADT modalities and
calculate the correlation between the genes using all cells. The nodes
denote proteins and RNA whilst the edges denote positive and negative
correlation in expression.</p>
<p><strong>Questions:</strong></p>
<ul>
<li>Do the results change if we use a different correlation method?</li>
<li>What if we vary the <code>cor_threshold</code> parameter?</li>
<li>How do we examine the network for a specific cluster?</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RNA_feature_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">rna_DEgenes_all</span>, <span class="st">"[["</span>, <span class="st">"name"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ADT_feature_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">adt_DEgenes_all</span>, <span class="st">"[["</span>, <span class="st">"name"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="va">netw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/geneADTnetwork.html" class="external-link">geneADTnetwork</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,</span>
<span>               RNA_exprs_value <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>               altExp_name <span class="op">=</span> <span class="st">"ADT"</span>,</span>
<span>               altExp_exprs_value <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>               cell_subset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>               cor_threshold <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>               RNA_feature_subset <span class="op">=</span> <span class="va">RNA_feature_subset</span>,</span>
<span>               ADT_feature_subset <span class="op">=</span> <span class="va">ADT_feature_subset</span>,</span>
<span>               cor_method <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>               network_layout <span class="op">=</span> <span class="fu">igraph</span><span class="fu">::</span><span class="va"><a href="https://r.igraph.org/reference/layout_with_fr.html" class="external-link">layout_with_fr</a></span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-21-1.png" width="1152"></p>
<p><strong>Interpretation:</strong></p>
<p>There are roughly 4 groupings in the interaction network,
particularly of interest, CD27 and CD11b is positively correlated with
RPS and RPL ribosomal genes. This could be expected, as activated immune
cells require increased ribosome biogenesis to support the immune
response.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="what-are-the-ligand-receptor-interactions-in-my-data">What are the ligand-receptor interactions in my data?<a class="anchor" aria-label="anchor" href="#what-are-the-ligand-receptor-interactions-in-my-data"></a>
</h3>
<p>Ligand-receptor interactions are the means by which cells communicate
with each other. Ligands (eg. hormones, neurotransmitters, growth
factors etc.) bind to specific receptors on the surface of target cells,
initiating intracellular signalling cascades that alter cell
behaviour.</p>
<p>To take advantage of the ADT expression, CiteFuse implements a
<code>ligandReceptorTest</code> to find ligand-receptor interactions
between sender and receiver cells. Importantly, the ADT count is used to
predict receptor expression within receiver cells. Note that the setting
<code>altExp_name = "RNA"</code> would enable users to predict
ligand-receptor interaction from RNA expression only.</p>
<p>First, we load a database of known ligand-receptor interactions to
query against our data.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the database of known ligand-receptor interactions from cellPhoneDB</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"lr_pair_subset"</span>, package <span class="op">=</span> <span class="st">"CiteFuse"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lr_pair_subset</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1]          [,2]   </span></span>
<span><span class="co">#&gt; [1,] "hg19_IL17RA" "CD45" </span></span>
<span><span class="co">#&gt; [2,] "hg19_FAS"    "CD11b"</span></span>
<span><span class="co">#&gt; [3,] "hg19_GZMK"   "CD62L"</span></span>
<span><span class="co">#&gt; [4,] "hg19_CD40LG" "CD11b"</span></span>
<span><span class="co">#&gt; [5,] "hg19_FLT3LG" "CD62L"</span></span>
<span><span class="co">#&gt; [6,] "hg19_GZMA"   "CD19"</span></span></code></pre></div>
<p>A key challenge in analyzing ligand–receptor interactions is the
difference in scaling and distribution. Comparing unequal distributions
can potentially introduce bias, especially during ligand–receptor
predictions when means are compared. To address this, we first scaled
each feature to a range of 0–1 through min–max normalization and then
transformed the protein expression to force the low-expression values to
zero.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html" class="external-link">normaliseExprs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">sce_citeseq</span>, </span>
<span>                              altExp_name <span class="op">=</span> <span class="st">"ADT"</span>, </span>
<span>                              transform <span class="op">=</span> <span class="st">"zi_minMax"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html" class="external-link">normaliseExprs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">sce_citeseq</span>, </span>
<span>                              altExp_name <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                              exprs_value <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                              transform <span class="op">=</span> <span class="st">"minMax"</span><span class="op">)</span></span></code></pre></div>
<p>For each ligand–receptor interaction pair originating from a cluster
expressing the ligand and another cluster expressing the receptor, a
permutation test is performed on the mean of the average RNA expression
from the ligand cluster and the mean of the protein expression from the
receptor cluster. Only ligand–receptor pairs with a P-value of lower
than 0.05 are defined as significant pairs.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/ligandReceptorTest.html" class="external-link">ligandReceptorTest</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">sce_citeseq</span>,</span>
<span>                                  ligandReceptor_list <span class="op">=</span> <span class="va">lr_pair_subset</span>,</span>
<span>                                  cluster <span class="op">=</span> <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_louvain</span>,</span>
<span>                                  RNA_exprs_value <span class="op">=</span> <span class="st">"minMax"</span>,</span>
<span>                                  use_alt_exp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  altExp_name <span class="op">=</span> <span class="st">"ADT"</span>,</span>
<span>                                  altExp_exprs_value <span class="op">=</span> <span class="st">"zi_minMax"</span>,</span>
<span>                                  num_permute <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span> </span>
<span><span class="co">#&gt; 100 ......200 ......300 ......400 ......500 ......600 ......700 ......800 ......900 ......1000 ......</span></span></code></pre></div>
<p><strong>Alternative visualisation 1:</strong> To visualise the
ligand-receptor predictions, we can plot a heatmap of p-values
indicating significant ligand-receptor pairs between cluster groups.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visLigandReceptor.html" class="external-link">visLigandReceptor</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>                  type <span class="op">=</span> <span class="st">"pval_heatmap"</span>,</span>
<span>                  receptor_type <span class="op">=</span> <span class="st">"ADT"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-25-1.png" width="768"></p>
<p><strong>Alternative visualisation 2:</strong> Here, we plot a network
graph between cluster groups, where the weight of the lines indicate the
number of LR-pairs.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visLigandReceptor.html" class="external-link">visLigandReceptor</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>                  type <span class="op">=</span> <span class="st">"group_network"</span>,</span>
<span>                  receptor_type <span class="op">=</span> <span class="st">"ADT"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-26-1.png" width="768"></p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="evaluating-the-importance-of-adt-markers-in-clustering-using-random-forest">Evaluating the importance of ADT markers in clustering using random
forest<a class="anchor" aria-label="anchor" href="#evaluating-the-importance-of-adt-markers-in-clustering-using-random-forest"></a>
</h2>
<p>An important evaluation in CITE-seq data analysis is to assess the
quality of each ADT and to evaluate the contribution of ADTs towards
clustering outcome. <code>CiteFuse</code> calculates the relative
importance of ADT towards clustering outcome by using a random forest
model. The higher the score of an ADT, the greater its importance
towards the final clustering outcome.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/importanceADT.html" class="external-link">importanceADT</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, </span>
<span>                             group <span class="op">=</span> <span class="va">sce_citeseq</span><span class="op">$</span><span class="va">SNF_W_louvain</span>,</span>
<span>                             subsample <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The importance scores can be visualised in a boxplot. Our evaluation
of ADT importance show that unsurprisingly CD4 and CD8 are the top two
discriminating proteins in PBMCs.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/visImportance.html" class="external-link">visImportance</a></span><span class="op">(</span><span class="va">sce_citeseq</span>, plot <span class="op">=</span> <span class="st">"boxplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="citefuse_files/figure-html/unnamed-chunk-28-1.png" width="1152"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># view the top 20 markers by decreasing importance</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce_citeseq</span><span class="op">)</span><span class="op">[[</span><span class="st">"importanceADT"</span><span class="op">]</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span></span>
<span><span class="co">#&gt;              CD27               CD8               CD4      PECAM (CD31) </span></span>
<span><span class="co">#&gt;         42.429076         37.959710         35.452469         16.071124 </span></span>
<span><span class="co">#&gt;               CD7             CD11b               CD5              CD28 </span></span>
<span><span class="co">#&gt;         14.188008         13.978868         13.292323         12.895289 </span></span>
<span><span class="co">#&gt;              CD44 IL7Ralpha (CD127)               CD2    MHCII (HLA-DR) </span></span>
<span><span class="co">#&gt;         10.896644         10.336379         10.194173          9.947160 </span></span>
<span><span class="co">#&gt;             CD11c         HLA-A,B,C      CD366 (tim3)             CD62L </span></span>
<span><span class="co">#&gt;          9.913175          7.044662          7.014305          6.759411 </span></span>
<span><span class="co">#&gt;            CD45RA               CD3              CD19      PD-1 (CD279) </span></span>
<span><span class="co">#&gt;          6.697095          5.245505          5.094099          4.112335</span></span></code></pre></div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Carissa Chen, Sharon Long, Pengyi Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
