<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SCMultiomeCNRSR">
<title>Introduction to Single-cell Pre-processing • SCMultiomeCNRSR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to Single-cell Pre-processing">
<meta property="og:description" content="SCMultiomeCNRSR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SCMultiomeCNRSR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction_to_single-cell.html">Introduction to single-cell pre-processing</a>
    <a class="dropdown-item" href="../articles/basics_of_single-cell_annotation.html">Basics of single-cell annotation</a>
    <a class="dropdown-item" href="../articles/advancedphenotyping.html">Advanced cell phenotyping</a>
    <a class="dropdown-item" href="../articles/citefuse.html">Analysing CITE-seq data with CiteFuse</a>
    <a class="dropdown-item" href="../articles/introduction_to_10x_multiome.html">Introduction to 10x multiome</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/carissaynchen/SCMultiome-CNRS-workshop-R">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to Single-cell Pre-processing</h1>
                        <h4 data-toc-skip class="author">Carissa Chen<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:cchen@cmri.org.au" class="email"&gt;cchen@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a> Sharon
Long<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:slong@cmri.org.au" class="email"&gt;slong@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>
Pengyi Yang<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:pyang@cmri.org.au" class="email"&gt;pyang@cmri.org.au&lt;/a&gt;&lt;/p&gt;'><sup>3</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/carissaynchen/SCMultiome-CNRS-workshop-R/blob/HEAD/vignettes/introduction_to_single-cell.Rmd" class="external-link"><code>vignettes/introduction_to_single-cell.Rmd</code></a></small>
      <div class="d-none name"><code>introduction_to_single-cell.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>Single-cell RNA-sequencing has allowed us to dissect cellular
heterogeneity at a finer resolution than previously possible with bulk
RNA-sequencing. However, the increase in dimensionality comes with its
unique challenges in data processing and analysis. In this section of
the workshop, we introduce the basic pipeline to check the quality of
single-cell RNA-seq processed data and further perform necessary
filtering to remove low-quality cells. The <em>learning goals</em> of
this workshop is to understand the implications of filtering and learn
how to detect low-quality cells.</p>
<p><br></p>
<p><span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-pencil-alt fa-stack-1x fa-inverse"></i></span>
<strong>Preparation and assumed knowledge</strong></p>
<ul>
<li>Knowledge of R syntax</li>
<li>Familiarity with the <a href="https://bioconductor.org/books/release/OSCA/index.html" class="external-link">Single-cell
data analysis guide</a>
</li>
<li>Familiarity with the <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html" class="external-link">SingleCellExperiment
class</a>
</li>
</ul>
<p><br></p>
<p><span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-location-arrow fa-stack-1x fa-inverse"></i></span>
<strong>Learning objectives</strong></p>
<ul>
<li>Carry out typical single-cell processing and QC workflow</li>
<li>generate appropriate QC graphics</li>
</ul>
<p><br></p>
</div>
<div class="section level3">
<h3 id="schedule">Schedule<a class="anchor" aria-label="anchor" href="#schedule"></a>
</h3>
<p>Structure of this section:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Data quality inspection</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Cell filtering and generating QC plots</td>
<td>20m</td>
</tr>
<tr class="even">
<td>Identifying cell types</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Concluding remarks</td>
<td></td>
</tr>
</tbody>
</table>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>scRNA-seq data is often plagued by various sources of noise and
technical artifacts, such as low-quality cells (eg. cells with high
rates of cell death as a result of single-cell dissociation), doublets,
and ambient RNA contamination.</p>
<p>Adequately filtering the data to retain good quality cells, while
minimising technical artifacts is crucial to ensure the accuracy and
reliability of downstream analyses. Without proper filtering, results
can be skewed by spurious signals, leading to inaccurate biological
interpretations and conclusions.</p>
<p>Typical procedures for scRNA-seq filtering include the removal of
cells with low or excessively high library sizes, cells with a high
proportion of mitochondrial gene expression (indicative of high cell
death) and removal of doublet cells (ie. when two cells are encapsulated
into one droplet, confounding what is considered ‘real cells’). These
steps help to retain high-quality data that accurately reflects the true
biological state of the sampled cells, thereby enabling more robust and
meaningful insights from the scRNA-seq experiment.</p>
<p>In this workshop, we will use a mouse scRNA-seq data profiling the
hippocampus from <a href="http://doi.org/10.1038/s41593-017-0056-2" class="external-link">Hochgerner et
al. (2018)</a>. The original data can be accessed from GEO with the
accession number GSE104323. The original data contains 8 timepoints from
E16.5 to postnatal 132 days, and 24 unique cell types.</p>
<p>While we do not have access to the count matrices directly after
running CellRanger, for the purposes of demonstration, we will assume
that the dataset is the raw, un-processed count matrix, where each age
is a separate sample.</p>
</div>
<div class="section level2">
<h2 id="loading-libraries-and-the-data">Loading libraries and the data<a class="anchor" aria-label="anchor" href="#loading-libraries-and-the-data"></a>
</h2>
<p>First, let us load common packages for single-cell processing.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jrnold.github.io/ggthemes/" class="external-link">ggthemes</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/plger/scDblFinder" class="external-link">scDblFinder</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DropletUtils</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We provide the original data downloaded from GSE104323 into a
SingleCellExperiment class object. It contains 24,185 cells and 27,932
genes. Below are typical functions to interact with the SCE object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># code to load data </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"../data/hippocampus_all_ages.rda"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">hippocampus_all_ages</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">hippocampus_all_ages</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 27932 24185</span></span>
<span><span class="co"># colData(sce) # view column metadata</span></span>
<span><span class="co"># rowData(sce) # view row metadata</span></span>
<span><span class="co"># counts(sce) # extract counts matrix, warning - very large</span></span>
<span><span class="co"># rownames(sce) # extract rownames and column names as usual</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="initial-inspection-of-the-data">Initial inspection of the data<a class="anchor" aria-label="anchor" href="#initial-inspection-of-the-data"></a>
</h2>
<div class="section level3">
<h3 id="examining-sequencing-depth">Examining sequencing depth<a class="anchor" aria-label="anchor" href="#examining-sequencing-depth"></a>
</h3>
<p>Assuming sequencing samples are separated by age, we can first look
at the reads obtained per library by aggregating the total cell counts
by age.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reads</span> <span class="op">&lt;-</span> <span class="fu">matrixStats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSums2.html" class="external-link">colSums2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">reads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">reads</span><span class="op">)</span>, <span class="va">sce</span><span class="op">$</span><span class="va">age</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">reads</span> <span class="op">&lt;-</span> <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">reads</span><span class="op">)</span></span>
<span><span class="va">reads</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">reads</span><span class="op">)</span></span>
<span><span class="va">reads</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">reads</span><span class="op">$</span><span class="va">sample</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"E16.5"</span>, <span class="st">"P0"</span>, <span class="st">"P5"</span>, <span class="st">"P18"</span>, <span class="st">"P19"</span>, <span class="st">"P23"</span>, <span class="st">"P120"</span>, <span class="st">"P132"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">reads</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">value</span>, x<span class="op">=</span><span class="va">sample</span>, fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_single-cell_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p><strong>Interpretation:</strong></p>
<p>From this bar plot, we see that each age has different, particularly
P120. Given the sequencing was not run in separate batches, we ideally
expect the total reads to be approximately similar. Therefore, at this
stage, we can potentially flag low quality cells for samples with low
total reads.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="generating-qc-metrics-using-scater">Generating QC metrics using <code>scater</code><a class="anchor" aria-label="anchor" href="#generating-qc-metrics-using-scater"></a>
</h3>
<p>Next, we use <code>addPerCellQC</code> function from the
<code>scater</code> package to calculate per-cell QC metrics for our
count matrix. This gives us the sum of counts (‘sum’), number of
detected features (‘detected’) and we also calculate the same metrics
for mitochondrial genes, a common method to measure cell death. We
measure the sparsity of gene expression in a cell by calculating the
percentage of zero counts per cell.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mito <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQC</a></span><span class="op">(</span><span class="va">sce</span>, assay.type <span class="op">=</span> <span class="st">"counts"</span>, subsets<span class="op">=</span><span class="va">subset</span>, threshold<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">pZero</span> <span class="op">&lt;-</span> <span class="fu">matrixStats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSums2.html" class="external-link">colSums2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">13</span><span class="op">:</span><span class="fl">18</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nUMI"</span>, <span class="st">"nGenes"</span>, <span class="st">"nMito"</span>, <span class="st">"mitoGenes_detected"</span>, <span class="st">"pMito"</span>, <span class="st">"total"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 19 columns</span></span>
<span><span class="co">#&gt;                                   sample_name   source_name     organism</span></span>
<span><span class="co">#&gt;                                   &lt;character&gt;   &lt;character&gt;  &lt;character&gt;</span></span>
<span><span class="co">#&gt; 10X79_1_TCTACCATGCCTAA 10X79_1_TCTACCATGCCT.. dentate gyrus Mus musculus</span></span>
<span><span class="co">#&gt; 10X79_2_GTACTAGTGAACAT 10X79_2_GTACTAGTGAAC.. dentate gyrus Mus musculus</span></span>
<span><span class="co">#&gt; 10X79_2_AATCAGTACCTACA 10X79_2_AATCAGTACCTA.. dentate gyrus Mus musculus</span></span>
<span><span class="co">#&gt; 10X79_1_CGGGTTCTTGAGGT 10X79_1_CGGGTTCTTGAG.. dentate gyrus Mus musculus</span></span>
<span><span class="co">#&gt; 10X79_1_GTGGAAGGCGTACA 10X79_1_GTGGAAGGCGTA.. dentate gyrus Mus musculus</span></span>
<span><span class="co">#&gt; 10X79_1_GTCCGCAAGCCATT 10X79_1_GTCCGCAAGCCA.. dentate gyrus Mus musculus</span></span>
<span><span class="co">#&gt;                             strain         age sex_of_pooled_animals</span></span>
<span><span class="co">#&gt;                        &lt;character&gt; &lt;character&gt;           &lt;character&gt;</span></span>
<span><span class="co">#&gt; 10X79_1_TCTACCATGCCTAA   hGFAP-GFP        P120        2males+1female</span></span>
<span><span class="co">#&gt; 10X79_2_GTACTAGTGAACAT     C57Bl/6         P19                female</span></span>
<span><span class="co">#&gt; 10X79_2_AATCAGTACCTACA     C57Bl/6         P19                female</span></span>
<span><span class="co">#&gt; 10X79_1_CGGGTTCTTGAGGT   hGFAP-GFP        P120        2males+1female</span></span>
<span><span class="co">#&gt; 10X79_1_GTGGAAGGCGTACA   hGFAP-GFP        P120        2males+1female</span></span>
<span><span class="co">#&gt; 10X79_1_GTCCGCAAGCCATT   hGFAP-GFP        P120        2males+1female</span></span>
<span><span class="co">#&gt;                        cell_cluster    molecule srr_run_accession</span></span>
<span><span class="co">#&gt;                         &lt;character&gt; &lt;character&gt;       &lt;character&gt;</span></span>
<span><span class="co">#&gt; 10X79_1_TCTACCATGCCTAA          RGL   total RNA        SRR6090459</span></span>
<span><span class="co">#&gt; 10X79_2_GTACTAGTGAACAT          RGL   total RNA        SRR6099081</span></span>
<span><span class="co">#&gt; 10X79_2_AATCAGTACCTACA          RGL   total RNA        SRR6099074</span></span>
<span><span class="co">#&gt; 10X79_1_CGGGTTCTTGAGGT          RGL   total RNA        SRR6090410</span></span>
<span><span class="co">#&gt; 10X79_1_GTGGAAGGCGTACA          RGL   total RNA        SRR6090448</span></span>
<span><span class="co">#&gt; 10X79_1_GTCCGCAAGCCATT          RGL   total RNA        SRR6090444</span></span>
<span><span class="co">#&gt;                        raw_file_original_file_name   umi_cellular_barcode</span></span>
<span><span class="co">#&gt;                                        &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">#&gt; 10X79_1_TCTACCATGCCTAA      10X79_1_TCTACCATGCCT.. TATGTCCCAG_TCTACCATG..</span></span>
<span><span class="co">#&gt; 10X79_2_GTACTAGTGAACAT      10X79_2_GTACTAGTGAAC.. CAGCTATTGG_GTACTAGTG..</span></span>
<span><span class="co">#&gt; 10X79_2_AATCAGTACCTACA      10X79_2_AATCAGTACCTA.. TTTAGCCTAT_AATCAGTAC..</span></span>
<span><span class="co">#&gt; 10X79_1_CGGGTTCTTGAGGT      10X79_1_CGGGTTCTTGAG.. CGTGAAGCGT_CGGGTTCTT..</span></span>
<span><span class="co">#&gt; 10X79_1_GTGGAAGGCGTACA      10X79_1_GTGGAAGGCGTA.. GGAGCCCGAC_GTGGAAGGC..</span></span>
<span><span class="co">#&gt; 10X79_1_GTCCGCAAGCCATT      10X79_1_GTCCGCAAGCCA.. ACAACCAGTC_GTCCGCAAG..</span></span>
<span><span class="co">#&gt;                                     cell_name      nUMI    nGenes     nMito</span></span>
<span><span class="co">#&gt;                                   &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 10X79_1_TCTACCATGCCTAA 10X79_1_TCTACCATGCCTAA      2287      1201         0</span></span>
<span><span class="co">#&gt; 10X79_2_GTACTAGTGAACAT 10X79_2_GTACTAGTGAACAT      1847      1128         0</span></span>
<span><span class="co">#&gt; 10X79_2_AATCAGTACCTACA 10X79_2_AATCAGTACCTACA      2216      1202         0</span></span>
<span><span class="co">#&gt; 10X79_1_CGGGTTCTTGAGGT 10X79_1_CGGGTTCTTGAGGT      1455       999         0</span></span>
<span><span class="co">#&gt; 10X79_1_GTGGAAGGCGTACA 10X79_1_GTGGAAGGCGTACA      1553       943         0</span></span>
<span><span class="co">#&gt; 10X79_1_GTCCGCAAGCCATT 10X79_1_GTCCGCAAGCCATT      2568      1534         0</span></span>
<span><span class="co">#&gt;                        mitoGenes_detected     pMito     total     pZero</span></span>
<span><span class="co">#&gt;                                 &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 10X79_1_TCTACCATGCCTAA                  0         0      2287  0.957003</span></span>
<span><span class="co">#&gt; 10X79_2_GTACTAGTGAACAT                  0         0      1847  0.959616</span></span>
<span><span class="co">#&gt; 10X79_2_AATCAGTACCTACA                  0         0      2216  0.956967</span></span>
<span><span class="co">#&gt; 10X79_1_CGGGTTCTTGAGGT                  0         0      1455  0.964235</span></span>
<span><span class="co">#&gt; 10X79_1_GTGGAAGGCGTACA                  0         0      1553  0.966239</span></span>
<span><span class="co">#&gt; 10X79_1_GTCCGCAAGCCATT                  0         0      2568  0.945081</span></span></code></pre></div>
<p>Then we visualise the per-cell QC metrics using a boxplot
visualisation.</p>
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<ul>
<li>What can you infer about the quality of the samples from these
boxplots?</li>
<li>How would you choose the appropriate filtering thresholds?</li>
<li>Should separate filtering thresholds be applied to each sample?</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nUMI"</span>, <span class="st">"nGenes"</span>, <span class="st">"pMito"</span>, <span class="st">"pZero"</span>, <span class="st">"age"</span><span class="op">)</span></span>
<span><span class="va">qc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span> , <span class="va">metrics</span><span class="op">]</span></span>
<span></span>
<span><span class="va">gg_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">qc_plot</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">age</span>,</span>
<span>                   metric <span class="op">=</span> <span class="va">qc_plot</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">metric_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">qc_plot</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">gg_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">metric</span>, <span class="va">median</span><span class="op">)</span>, y<span class="op">=</span><span class="va">metric</span>, fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">metric_name</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">gg_list</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, ncol<span class="op">=</span><span class="fl">2</span>, nrow<span class="op">=</span><span class="fl">2</span>, guides<span class="op">=</span><span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_single-cell_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="performing-cell-filtering">Performing cell filtering<a class="anchor" aria-label="anchor" href="#performing-cell-filtering"></a>
</h2>
<p>Now that we have visualised the distribution of our QC metrics using
a boxplot, it becomes clearer which thresholds we could apply to remove
outlier cells. Generally, we try to remove cells that have very low UMI
counts and very sparse cells with a high percentage of zero counts as
they usually indicate cells with insufficient sequencing depth. We also
remove cells with extremely high UMI counts as that may indicate the
presence of doublets. Lastly, we usually set a threshold according to
mitochondrial gene expression to remove cells that have high rates of
cell death occurring.</p>
<p>It is important to keep in mind that all these cut-offs are
dataset-dependent and in some cases, you may want to set lower or more
stringent thresholds to avoid removing rare cell populations or to
capture cell types which may have specific characteristics. A study by
<a href="https://doi.org/10.1093/bioinformatics/btaa751" class="external-link">Orsorio et
al.</a> has performed a systematic evaluation of mitochondrial
proportion in human and mice tissues which could serve as a guide to
decide the appropriate mitochondrial threshold depending on your
data.</p>
<p>In this step, we visualise the distribution of the QC metrics after
performing cell filtering.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nUMI"</span>, <span class="st">"nGenes"</span>, <span class="st">"pMito"</span>, <span class="st">"pZero"</span>, <span class="st">"age"</span><span class="op">)</span></span>
<span><span class="va">qc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span> , <span class="va">metrics</span><span class="op">]</span></span>
<span></span>
<span><span class="va">idx</span> <span class="op">=</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">nUMI</span> <span class="op">&lt;</span> <span class="fl">20000</span> <span class="op">&amp;</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">pMito</span> <span class="op">&lt;</span> <span class="fl">10</span> <span class="op">&amp;</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">pZero</span> <span class="op">&lt;</span> <span class="fl">0.98</span></span>
<span><span class="va">qc_plot</span> <span class="op">=</span> <span class="va">qc_plot</span><span class="op">[</span><span class="va">idx</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">filt_gg</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">qc_plot</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">age</span>,</span>
<span>                   metric <span class="op">=</span> <span class="va">qc_plot</span><span class="op">[</span>, <span class="va">x</span><span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span>   <span class="va">metric_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">qc_plot</span><span class="op">)</span><span class="op">[</span><span class="va">x</span><span class="op">]</span></span>
<span>   </span>
<span>    <span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">metric</span>, <span class="va">median</span><span class="op">)</span>, y<span class="op">=</span><span class="va">metric</span>, fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">metric_name</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">filt_gg</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, ncol<span class="op">=</span><span class="fl">2</span>, nrow<span class="op">=</span><span class="fl">2</span>, guides<span class="op">=</span><span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_single-cell_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>We can also visualise how many cells are lost/remaining after
applying the filtering thresholds. Based on these visualisations, which
filtering criteria do you think has the most impact on the number of
cells retained?</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nUMI"</span>, <span class="st">"nGenes"</span>, <span class="st">"pMito"</span>, <span class="st">"pZero"</span>, <span class="st">"age"</span><span class="op">)</span></span>
<span><span class="va">qc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span> , <span class="va">metrics</span><span class="op">]</span></span>
<span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">nUMI</span> <span class="op">&lt;</span> <span class="fl">20000</span> <span class="op">&amp;</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">pMito</span> <span class="op">&lt;</span> <span class="fl">10</span> <span class="op">&amp;</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">pZero</span> <span class="op">&lt;</span> <span class="fl">0.98</span></span>
<span><span class="va">qc_plot</span><span class="op">$</span><span class="va">QC_cells</span> <span class="op">&lt;-</span> <span class="va">idx</span></span>
<span><span class="va">qc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">qc_plot</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">qc_plot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">age</span>, fill <span class="op">=</span> <span class="va">QC_cells</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>position<span class="op">=</span><span class="st">"fill"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_single-cell_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="performing-doublet-detection-using-scdblfinder">Performing doublet detection using <code>scDblFinder</code><a class="anchor" aria-label="anchor" href="#performing-doublet-detection-using-scdblfinder"></a>
</h2>
<p>Single-cell RNA-seq experimental protocols require a cell suspension
to be distributed across single droplets (droplet microfluidics) or
single wells (well-based) with the intention of capturing one cell per
reaction volume (ie. a singlet). Then, mRNA molecules are labeled with a
unique molecular code (UMI barcode). However, during the distribution
step, one droplet or well may capture more than one cell, creating a
‘doublet’ cell that is then sequenced as a single cell. The doublet rate
(i.e., the proportion of doublets) in a scRNA-seq experiment depends on
the throughput and protocol. There are two major classes of doublets:
homotypic doublets, which are formed by transcriptionally similar cells;
and heterotypic doublets, which are formed by cells of distinct types,
lineages, or states.</p>
<p>Hence, doublet detection and removal is an important step in
pre-processing as it can bias the interpretation of downstream analyses
such as differential gene expression analysis, cell clustering and
trajectory analysis. Here, we use <code>scDblFinder</code> which has
been shown to be <a href="https://github.com/plger/scDblFinder" class="external-link">competitive</a> with other
top-performing methods according to this excellent benchmark by <a href="https://doi.org/10.1016/j.cels.2020.11.008" class="external-link">Xi et al.</a>.</p>
<p>After running <code>scDblFinder</code>, we first plot a bar plot to
visualise how many doublets we have in our dataset. Then, we can
additionally confirm they are indeed doublets by visualisation their
distribution of UMI counts, number of genes and proportion of zero
expression compared to singlets.</p>
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<ul>
<li>What do you notice about the differences in distribution between the
singlets and doublets?</li>
<li>Is this expected and why?</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scDblFinder</span><span class="fu">::</span><span class="fu"><a href="https://plger.github.io/scDblFinder/reference/scDblFinder.html" class="external-link">scDblFinder</a></span><span class="op">(</span><span class="va">sce</span>, samples<span class="op">=</span><span class="st">"age"</span>, clusters<span class="op">=</span><span class="cn">TRUE</span>, BPPARAM<span class="op">=</span><span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">age</span>, doublet <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">scDblFinder.class</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">qc_plot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">sample</span>,fill <span class="op">=</span> <span class="va">doublet</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>position<span class="op">=</span><span class="st">"dodge"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_single-cell_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nUMI"</span>, <span class="st">"nGenes"</span>, <span class="st">"pMito"</span>, <span class="st">"pZero"</span>, <span class="st">"age"</span>, <span class="st">"scDblFinder.class"</span><span class="op">)</span></span>
<span><span class="va">qc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span> , <span class="va">metrics</span><span class="op">]</span></span>
<span></span>
<span><span class="va">idx</span> <span class="op">=</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">nUMI</span> <span class="op">&lt;</span> <span class="fl">20000</span> <span class="op">&amp;</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">pMito</span> <span class="op">&lt;</span> <span class="fl">10</span> <span class="op">&amp;</span> <span class="va">qc_plot</span><span class="op">$</span><span class="va">pZero</span> <span class="op">&lt;</span> <span class="fl">0.98</span></span>
<span><span class="va">qc_plot</span><span class="op">$</span><span class="va">QC_cells</span> <span class="op">&lt;-</span> <span class="va">idx</span></span>
<span></span>
<span><span class="va">gg_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">qc_plot</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>metric <span class="op">=</span> <span class="va">qc_plot</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span>,</span>
<span>                   sample <span class="op">=</span> <span class="va">qc_plot</span><span class="op">[</span>, <span class="st">"age"</span><span class="op">]</span>,</span>
<span>                   doublet_class <span class="op">=</span> <span class="va">qc_plot</span><span class="op">[</span>, <span class="st">"scDblFinder.class"</span><span class="op">]</span>,</span>
<span>                   QC_cells <span class="op">=</span> <span class="va">qc_plot</span><span class="op">[</span>, <span class="st">"QC_cells"</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">$</span><span class="va">QC_cells</span> <span class="op">==</span> <span class="cn">TRUE</span>, <span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">metric_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">qc_plot</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>  <span class="va">gg_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">metric</span>, <span class="va">median</span><span class="op">)</span>, y<span class="op">=</span><span class="va">metric</span>, fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">doublet_class</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">metric_name</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle<span class="op">=</span><span class="fl">45</span>, vjust<span class="op">=</span><span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">gg_list</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span>, ncol<span class="op">=</span><span class="fl">2</span>, nrow<span class="op">=</span><span class="fl">2</span>, guides<span class="op">=</span><span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_single-cell_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p><strong>Interpretation:</strong></p>
<p>From the boxplots, we see that the doublet population generally has
higher total UMI count, number of genes expressed and lower proportion
of zero counts. This in line with our expectations, because it is has
predicted the combined transcriptomic expression from two cells instead
of a single cell.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="save-the-processed-data">Save the processed data<a class="anchor" aria-label="anchor" href="#save-the-processed-data"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># idx = sce$nUMI &lt; 20000 &amp; sce$pMito &lt; 10 &amp; sce$pZero &lt; 0.98 &amp; sce$scDblFinder.class == "singlet"</span></span>
<span><span class="co"># sce &lt;- sce[, idx]</span></span>
<span></span>
<span><span class="co"># save(sce, file="data/hippocampus_processed.rda")</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="extension-filtering-empty-droplets-using-dropletutils">Extension – filtering empty droplets using
<code>DropletUtils</code><a class="anchor" aria-label="anchor" href="#extension-filtering-empty-droplets-using-dropletutils"></a>
</h2>
<p>Empty droplets often contain RNA from the ambient solution, resulting
in non-zero counts after debarcoding. Thus, to remove the experimental
artifacts, we want to distinguish between empty droplets and real cells.
the <code>emptyDrops</code> function testing each barcode’s expression
profile for significant deviation from the ambient profile.</p>
<p>Again, as we are not working with the raw CellRanger output, we will
simulate a count matrix with empty droplets to demonstrate the utility
of this step. Note, this step is typically run on the <em>raw barcode
matrix</em> from CellRanger, where empty droplets have not been removed
by default. While it is perfectly fine to use the <em>filtered barcode
matrix</em>, depending on situation, it may be desirable to start from
the unfiltered matrix to preserve more cells.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DropletUtils</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">my.counts</span> <span class="op">&lt;-</span> <span class="fu">DropletUtils</span><span class="fu">:::</span><span class="fu">simCounts</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>We compute the statistics using the <code>barcodeRanks</code>
function, and then create the plot as shown below. This plot shows the
log-total count against the log-rank of each barcode, where the highest
ranked barcodes have the lowest total UMI counts. It also shows the
inflection and knee points on the curve, corresponding to the division
of the total count distribution, predicting the difference between empty
droplets and cell-containing droplets.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">br.out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/barcodeRanks.html" class="external-link">barcodeRanks</a></span><span class="op">(</span><span class="va">my.counts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Making a plot.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">br.out</span><span class="op">$</span><span class="va">rank</span>, <span class="va">br.out</span><span class="op">$</span><span class="va">total</span>, log<span class="op">=</span><span class="st">"xy"</span>, xlab<span class="op">=</span><span class="st">"Rank"</span>, ylab<span class="op">=</span><span class="st">"Total"</span><span class="op">)</span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">br.out</span><span class="op">$</span><span class="va">rank</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="va">br.out</span><span class="op">$</span><span class="va">rank</span><span class="op">[</span><span class="va">o</span><span class="op">]</span>, <span class="va">br.out</span><span class="op">$</span><span class="va">fitted</span><span class="op">[</span><span class="va">o</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fu">metadata</span><span class="op">(</span><span class="va">br.out</span><span class="op">)</span><span class="op">$</span><span class="va">knee</span>, col<span class="op">=</span><span class="st">"dodgerblue"</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fu">metadata</span><span class="op">(</span><span class="va">br.out</span><span class="op">)</span><span class="op">$</span><span class="va">inflection</span>, col<span class="op">=</span><span class="st">"forestgreen"</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"bottomleft"</span>, lty<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dodgerblue"</span>, <span class="st">"forestgreen"</span><span class="op">)</span>, </span>
<span>    legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"knee"</span>, <span class="st">"inflection"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_single-cell_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Droplets with significant deviations from the ambient profile are
detected at a specified FDR threshold, e.g., with FDR below 1%. These
can be considered to be cell-containing droplets, with a frequency of
false positives (i.e., empty droplets) at the specified FDR.
Furthermore, droplets with very large counts are automatically retained
by setting their p-values to zero. This avoids discarding droplets
containing cells that are very similar to the ambient profile.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">e.out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/emptyDrops.html" class="external-link">emptyDrops</a></span><span class="op">(</span><span class="va">my.counts</span><span class="op">)</span></span>
<span><span class="va">e.out</span></span>
<span><span class="co">#&gt; DataFrame with 11100 rows and 5 columns</span></span>
<span><span class="co">#&gt;           Total   LogProb    PValue   Limited        FDR</span></span>
<span><span class="co">#&gt;       &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;logical&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1             2        NA        NA        NA         NA</span></span>
<span><span class="co">#&gt; 2             9        NA        NA        NA         NA</span></span>
<span><span class="co">#&gt; 3            20        NA        NA        NA         NA</span></span>
<span><span class="co">#&gt; 4            20        NA        NA        NA         NA</span></span>
<span><span class="co">#&gt; 5             1        NA        NA        NA         NA</span></span>
<span><span class="co">#&gt; ...         ...       ...       ...       ...        ...</span></span>
<span><span class="co">#&gt; 11096       215  -246.428 9.999e-05      TRUE 0.00013799</span></span>
<span><span class="co">#&gt; 11097       201  -250.234 9.999e-05      TRUE 0.00013799</span></span>
<span><span class="co">#&gt; 11098       247  -275.905 9.999e-05      TRUE 0.00013799</span></span>
<span><span class="co">#&gt; 11099       191  -228.763 9.999e-05      TRUE 0.00013799</span></span>
<span><span class="co">#&gt; 11100       198  -233.043 9.999e-05      TRUE 0.00013799</span></span>
<span></span>
<span><span class="va">is.cell</span> <span class="op">&lt;-</span> <span class="va">e.out</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;=</span> <span class="fl">0.01</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">is.cell</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 943</span></span></code></pre></div>
<p>Lastly, we can plot a diagnostic plot such as the total UMI counts
against the negative log-probability. Droplets detected as real cells
should have a large negative log-probabilities or very large total
counts.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">e.out</span><span class="op">$</span><span class="va">Total</span>, <span class="op">-</span><span class="va">e.out</span><span class="op">$</span><span class="va">LogProb</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">is.cell</span>, <span class="st">"red"</span>, <span class="st">"black"</span><span class="op">)</span>,</span>
<span>    xlab<span class="op">=</span><span class="st">"Total UMI count"</span>, ylab<span class="op">=</span><span class="st">"-Log Probability"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_single-cell_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Carissa Chen, Sharon Long, Pengyi Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
